{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/bin/python\r\n"
     ]
    }
   ],
   "source": [
    "!which python"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/usr/bin/sh: line 0: source: activate: file not found\r\n"
     ]
    }
   ],
   "source": [
    "!source activate py36"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "#sys.path.append('/home/faske/g/anaconda3/envs/py34/lib/python3.4/site-packages')\n",
    "sys.path.append('/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages')\n",
    "sys.path.append(\"/data/gpfs/assoc/parchmanlab/tfaske/ipynb/include_utils\")\n",
    "\n",
    "import ipyparallel as ipp\n",
    "import os, time\n",
    "import include_utils as u\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import random\n",
    "import numbers\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.patches as mpatches\n",
    "import matplotlib.cm as cm\n",
    "import matplotlib.colors as mcolors\n",
    "#import vcf\n",
    "from sklearn import preprocessing\n",
    "from subprocess import Popen, PIPE, call, check_output\n",
    "import seaborn as sns\n",
    "from IPython.display import FileLink\n",
    "import urllib.request as urllib2\n",
    "import dill\n",
    "import traceback\n",
    "from pandas import Series, DataFrame\n",
    "import gzip\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore',category=pd.io.pytables.PerformanceWarning)\n",
    "%config InlineBackend.figure_format = 'retina'\n",
    "from Bio import SeqIO\n",
    "#import pysam\n",
    "from collections import OrderedDict, namedtuple, Counter\n",
    "import operator\n",
    "import multiprocessing as mp\n",
    "import shutil\n",
    "import tempfile\n",
    "#from ipyparallel import Client\n",
    "import scandir\n",
    "import glob\n",
    "from Bio.SeqIO.QualityIO import FastqGeneralIterator\n",
    "import pickle\n",
    "import re\n",
    "from itertools import chain\n",
    "#import Levenshtein as lv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcfutils = \"vcfutils.pl\"\n",
    "vcftools = \"vcftools\"\n",
    "bcftools = \"bcftools\"\n",
    "bgzip = \"bgzip\"\n",
    "tabix = \"tabix\"\n",
    "java = \"java\"\n",
    "plink = \"plink --noweb\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "analysis_dir = '/data/gpfs/home/tfaske/d/frogs/ddocent/filtering'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/denovo/tfaske/frogs/ddocent/filtering\n"
     ]
    }
   ],
   "source": [
    "cd $analysis_dir"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/home/tfaske/d/frogs/ddocent/filtering/TotalRawSNPs.vcf.gz'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_file = os.path.join(analysis_dir, \"TotalRawSNPs.vcf.gz\")\n",
    "assert os.path.exists(vcf_file)\n",
    "vcf_file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## keep only biallelic first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/TotalRawSNPs.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--out RALU.biallelic\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 323 out of 323 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 395007 out of a possible 496627 Sites\n",
      "Run Time = 414.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --remove-indels \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_file \\\n",
    "--out $'RALU.biallelic'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_biallelic = os.path.join(analysis_dir, \"RALU.biallelic.recode.vcf\")\n",
    "vcf_biallelic_gz = vcf_biallelic + '.gz'\n",
    "!$bgzip -c $vcf_biallelic > {vcf_biallelic_gz}\n",
    "!$tabix {vcf_biallelic_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Remove by MAF, missing, and thin next"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/RALU.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.6\n",
      "\t--out RALU_miss60_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 323 out of 323 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 22106 out of a possible 395007 Sites\n",
      "Run Time = 51.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.6 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'RALU_miss60_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/RALU.biallelic.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--maf 0.02\n",
      "\t--thin 100\n",
      "\t--max-missing 0.7\n",
      "\t--out RALU_miss70_thin100_MAF2\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 323 out of 323 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 20784 out of a possible 395007 Sites\n",
      "Run Time = 49.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools \\\n",
    "--max-missing 0.7 \\\n",
    "--maf 0.02 \\\n",
    "--thin 100 \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--gzvcf \\\n",
    "$vcf_biallelic_gz \\\n",
    "--out $'RALU_miss70_thin100_MAF2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_filtered = \"RALU_miss70_thin100_MAF2.recode.vcf\"\n",
    "vcf_filtered_gz = \"%s.gz\" % vcf_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "!$bgzip -c $vcf_filtered > {vcf_filtered_gz}\n",
    "!$tabix {vcf_filtered_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove bad_indv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf RALU_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--missing-indv\n",
      "\t--out RALU_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 323 out of 323 Individuals\n",
      "Outputting Individual Missingness\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz --out $vcf_filtered_gz --missing-indv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_imiss(vcf_file):\n",
    "    imiss_file = !ls {vcf_file}.imiss\n",
    "    imiss_df = pd.read_csv(imiss_file[0], sep=\"\\t\")\n",
    "    imiss_df.index = imiss_df.INDV\n",
    "    return imiss_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPES_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>INDV</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>RL_AD1C_200</th>\n",
       "      <td>RL_AD1C_200</td>\n",
       "      <td>20784</td>\n",
       "      <td>0</td>\n",
       "      <td>1286</td>\n",
       "      <td>0.061874</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RL_AD1C_201</th>\n",
       "      <td>RL_AD1C_201</td>\n",
       "      <td>20784</td>\n",
       "      <td>0</td>\n",
       "      <td>433</td>\n",
       "      <td>0.020833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RL_AD1C_202</th>\n",
       "      <td>RL_AD1C_202</td>\n",
       "      <td>20784</td>\n",
       "      <td>0</td>\n",
       "      <td>392</td>\n",
       "      <td>0.018861</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RL_AD2C_203</th>\n",
       "      <td>RL_AD2C_203</td>\n",
       "      <td>20784</td>\n",
       "      <td>0</td>\n",
       "      <td>1356</td>\n",
       "      <td>0.065242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RL_AD2C_204</th>\n",
       "      <td>RL_AD2C_204</td>\n",
       "      <td>20784</td>\n",
       "      <td>0</td>\n",
       "      <td>400</td>\n",
       "      <td>0.019246</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    INDV  N_DATA  N_GENOTYPES_FILTERED  N_MISS    F_MISS\n",
       "INDV                                                                    \n",
       "RL_AD1C_200  RL_AD1C_200   20784                     0    1286  0.061874\n",
       "RL_AD1C_201  RL_AD1C_201   20784                     0     433  0.020833\n",
       "RL_AD1C_202  RL_AD1C_202   20784                     0     392  0.018861\n",
       "RL_AD2C_203  RL_AD2C_203   20784                     0    1356  0.065242\n",
       "RL_AD2C_204  RL_AD2C_204   20784                     0     400  0.019246"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df = get_imiss(vcf_filtered_gz)\n",
    "imiss_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    323.000000\n",
       "mean       0.150702\n",
       "std        0.253932\n",
       "min        0.009430\n",
       "25%        0.025091\n",
       "50%        0.046959\n",
       "75%        0.131183\n",
       "max        0.999326\n",
       "Name: F_MISS, dtype: float64"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "imiss_df.F_MISS.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(323, 26, 29, 46)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(imiss_df),len(imiss_df[imiss_df.F_MISS >= .6]),len(imiss_df[imiss_df.F_MISS >= .5]), len(imiss_df[imiss_df.F_MISS >= .25])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "bad_indv = imiss_df.INDV[imiss_df.F_MISS >= .6]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"bad_indv.txt\"), \"w\") as o:\n",
    "    o.write(\"INDV\\n\")\n",
    "    for elem in bad_indv.index:\n",
    "        o.write(\"%s\\n\" % elem)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf RALU_miss70_thin100_MAF2.recode.vcf.gz\n",
      "\t--remove /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/bad_indv.txt\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "Excluding individuals in 'exclude' list\n",
      "After filtering, kept 297 out of 323 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 21.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_filtered_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--remove {os.path.join(analysis_dir, \"bad_indv.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"snps_indv_removed\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_bad_remove = os.path.join(analysis_dir, \"snps_indv_removed.recode.vcf\")\n",
    "vcf_bad_remove_gz = vcf_bad_remove + \".gz\"\n",
    "!$bgzip -c {vcf_bad_remove} > {vcf_bad_remove_gz}\n",
    "!$tabix {vcf_bad_remove_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter snps further \n",
    "this needs to be done after removing individuals\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_vcf_stats(vcf_gz):\n",
    "    \n",
    "    stats = ['depth',\n",
    "            'site-depth',\n",
    "            'site-mean-depth',\n",
    "            'site-quality',\n",
    "            'missing-site',\n",
    "            'freq',\n",
    "            'counts',\n",
    "            'hardy',\n",
    "            'het']\n",
    "    \n",
    "    for stat in stats:\n",
    "        !$vcftools --gzvcf $vcf_gz \\\n",
    "        --out $vcf_gz \\\n",
    "        {\"--%s\" % stat} "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Depth for Each Site\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-mean-depth\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Depth for Each Site\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--site-quality\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Quality for Each Site\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 1.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--missing-site\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Site Missingness\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--freq\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Frequency Statistics...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--counts\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Frequency Statistics...\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--hardy\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting HWE statistics (but only for biallelic loci)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n",
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--het\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Individual Heterozygosity\n",
      "After filtering, kept 20784 out of a possible 20784 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "get_vcf_stats(vcf_bad_remove_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pd.set_option('display.max_columns', 100)\n",
    "\n",
    "def get_MAF(row):\n",
    "    try:\n",
    "        return np.min([row.A1_freq, row.A2_freq])\n",
    "    except:\n",
    "        print(row)\n",
    "        \n",
    "def get_correction(n):\n",
    "    #for finite sample size\n",
    "    return (2*n)/(2*n-1)\n",
    "\n",
    "def calculate_Fis(vals):\n",
    "    try:\n",
    "        data = [float(x) for x in vals.split(\"/\")]\n",
    "        assert len(data) == 3\n",
    "        num_individuals = np.sum(data)\n",
    "        total_alleles = 2*num_individuals\n",
    "        a1_count = 2*data[0]\n",
    "        a2_count = 2*data[2]\n",
    "        het_count = data[1]\n",
    "        a1_count += het_count\n",
    "        a2_count += het_count\n",
    "        a1_freq = a1_count/total_alleles\n",
    "        a2_freq = a2_count/total_alleles\n",
    "        assert a1_freq + a2_freq == 1.0\n",
    "        He = 2 * a1_freq * a2_freq * get_correction(num_individuals)\n",
    "        Ho = het_count/num_individuals\n",
    "        Fis = 1 - (Ho/He)\n",
    "        return Fis\n",
    "    except:\n",
    "        return -9\n",
    "\n",
    "def combine_vcf_stats(filedir, prefix):\n",
    "    \n",
    "    hardy_files = !ls {filedir}/{prefix}.hwe\n",
    "    hardy = pd.read_csv(hardy_files[0], sep=\"\\t\")\n",
    "\n",
    "    hardy.columns = ['CHROM', 'POS', 'OBS(HOM1/HET/HOM2)', 'E(HOM1/HET/HOM2)', 'ChiSq_HWE',\n",
    "       'P_HWE', 'P_HET_DEFICIT', 'P_HET_EXCESS']\n",
    "    hardy.index = hardy.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_files = !ls {filedir}/{prefix}.l* | grep -v log\n",
    "    loci_df = pd.concat([pd.read_csv(x, sep=\"\\t\", skiprows=0) for x in loci_files], axis=1)\n",
    "    chrom_pos = loci_df.ix[:,0:2]\n",
    "    \n",
    "    frq_files = !ls {filedir}/{prefix}.frq* | grep -v count\n",
    "    frq_data = []\n",
    "    h = open(frq_files[0])\n",
    "    header = h.readline().strip().split()\n",
    "    for line in h:\n",
    "        frq_data.append(line.strip().split('\\t'))\n",
    "\n",
    "    header = ['CHROM', 'POS', 'N_ALLELES', 'N_CHR', 'A1_FREQ', \"A2_FREQ\"]\n",
    "    frq_df = pd.DataFrame(frq_data)\n",
    "    print(frq_df.columns)\n",
    "    #frq_df = frq_df.drop([6,7],axis=1)\n",
    "    frq_df.columns = header\n",
    "    frq_df.index = frq_df.apply(lambda x: \"%s-%s\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = loci_df.drop(['CHROM','CHR','POS'], axis=1)\n",
    "    loci_df = pd.concat([chrom_pos, loci_df], axis=1)\n",
    "    loci_df.index = loci_df.apply(lambda x: \"%s-%d\" % (x.CHROM, x.POS), axis=1)\n",
    "    \n",
    "    loci_df = pd.concat([loci_df, frq_df, hardy], axis=1)\n",
    "    loci_df[\"A1_allele\"] = loci_df.apply(lambda row: row.A1_FREQ.split(\":\")[0], axis=1)\n",
    "    loci_df[\"A2_allele\"] = loci_df.apply(lambda row: row.A2_FREQ.split(\":\")[0], axis=1)\n",
    "    \n",
    "    loci_df[\"A1_freq\"] = loci_df.apply(lambda row: float(row.A1_FREQ.split(\":\")[1]), axis=1)\n",
    "    loci_df[\"A2_freq\"] = loci_df.apply(lambda row: float(row.A2_FREQ.split(\":\")[1]), axis=1)\n",
    "    \n",
    "    loci_df['MAF'] = loci_df.apply(get_MAF, axis=1)\n",
    "    loci_df = loci_df.drop(['CHROM', 'POS'], axis=1)\n",
    "    \n",
    "    loci_df['Fis'] = loci_df['OBS(HOM1/HET/HOM2)'].apply(calculate_Fis)\n",
    "    \n",
    "    return loci_df, frq_df, hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz'"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vcf_bad_remove_gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/ipykernel_launcher.py:45: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "/data/gpfs/assoc/parchmanlab/tfaske/anaconda3/envs/py36/lib/python3.6/site-packages/pandas/core/indexing.py:822: FutureWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#ix-indexer-is-deprecated\n",
      "  retval = getattr(retval, self.name)._getitem_axis(key, axis=i)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RangeIndex(start=0, stop=6, step=1)\n"
     ]
    }
   ],
   "source": [
    "loci_df, frq_df, hardy = combine_vcf_stats(analysis_dir,'snps_indv_removed.recode.vcf.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SUM_DEPTH</th>\n",
       "      <th>SUMSQ_DEPTH</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "      <th>VAR_DEPTH</th>\n",
       "      <th>N_DATA</th>\n",
       "      <th>N_GENOTYPE_FILTERED</th>\n",
       "      <th>N_MISS</th>\n",
       "      <th>F_MISS</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>...</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "      <th>A1_allele</th>\n",
       "      <th>A2_allele</th>\n",
       "      <th>A1_freq</th>\n",
       "      <th>A2_freq</th>\n",
       "      <th>MAF</th>\n",
       "      <th>Fis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_109-16</th>\n",
       "      <td>305305</td>\n",
       "      <td>450193771</td>\n",
       "      <td>1027.96000</td>\n",
       "      <td>460647.00000</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>37573.500000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>19.035410</td>\n",
       "      <td>3.396307e-07</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.478008e-07</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.797980</td>\n",
       "      <td>0.202020</td>\n",
       "      <td>0.202020</td>\n",
       "      <td>-0.251055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_156-11</th>\n",
       "      <td>328000</td>\n",
       "      <td>472002036</td>\n",
       "      <td>1104.38000</td>\n",
       "      <td>370832.00000</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>301506.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>223.728000</td>\n",
       "      <td>2.161918e-62</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.027941e-62</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.535354</td>\n",
       "      <td>0.464646</td>\n",
       "      <td>0.464646</td>\n",
       "      <td>-0.864780</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_160-14</th>\n",
       "      <td>236545</td>\n",
       "      <td>249594353</td>\n",
       "      <td>796.44800</td>\n",
       "      <td>206752.00000</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>313307.000000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>232.993500</td>\n",
       "      <td>4.080681e-65</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.064197e-65</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>0.530303</td>\n",
       "      <td>0.469697</td>\n",
       "      <td>0.469697</td>\n",
       "      <td>-0.882540</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_212-78</th>\n",
       "      <td>294044</td>\n",
       "      <td>383862802</td>\n",
       "      <td>990.04700</td>\n",
       "      <td>313329.00000</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2530.050000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.360573</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>7.183592e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.966330</td>\n",
       "      <td>0.033670</td>\n",
       "      <td>0.033670</td>\n",
       "      <td>-0.033101</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_220-41</th>\n",
       "      <td>158221</td>\n",
       "      <td>111715447</td>\n",
       "      <td>532.73100</td>\n",
       "      <td>92656.30000</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3058.980000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>1.027682</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.902908e-01</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>0.944444</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>0.055556</td>\n",
       "      <td>-0.057041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396255-35</th>\n",
       "      <td>830</td>\n",
       "      <td>4432</td>\n",
       "      <td>3.50211</td>\n",
       "      <td>6.46292</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>120</td>\n",
       "      <td>0.202020</td>\n",
       "      <td>0.047948</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>140.154700</td>\n",
       "      <td>8.440953e-11</td>\n",
       "      <td>8.440953e-11</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>0.962025</td>\n",
       "      <td>0.037975</td>\n",
       "      <td>0.037975</td>\n",
       "      <td>0.769493</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396263-18</th>\n",
       "      <td>1041</td>\n",
       "      <td>5737</td>\n",
       "      <td>3.81319</td>\n",
       "      <td>6.49806</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>48</td>\n",
       "      <td>0.080808</td>\n",
       "      <td>2370.400000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.411256</td>\n",
       "      <td>7.021218e-01</td>\n",
       "      <td>7.867396e-01</td>\n",
       "      <td>3.441632e-01</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.802198</td>\n",
       "      <td>0.197802</td>\n",
       "      <td>0.197802</td>\n",
       "      <td>-0.036910</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396286-6</th>\n",
       "      <td>5349</td>\n",
       "      <td>111427</td>\n",
       "      <td>18.01010</td>\n",
       "      <td>50.98300</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1768.980000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>8.282448</td>\n",
       "      <td>1.436292e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.767035e-04</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.856902</td>\n",
       "      <td>0.143098</td>\n",
       "      <td>0.143098</td>\n",
       "      <td>-0.165029</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396287-2</th>\n",
       "      <td>1553</td>\n",
       "      <td>18101</td>\n",
       "      <td>6.21200</td>\n",
       "      <td>33.95090</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>94</td>\n",
       "      <td>0.158249</td>\n",
       "      <td>300.164000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2.213322</td>\n",
       "      <td>2.308508e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.385372e-01</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.914000</td>\n",
       "      <td>0.086000</td>\n",
       "      <td>0.086000</td>\n",
       "      <td>-0.091904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396299-19</th>\n",
       "      <td>2481</td>\n",
       "      <td>25667</td>\n",
       "      <td>8.38176</td>\n",
       "      <td>16.51480</td>\n",
       "      <td>594</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0.003367</td>\n",
       "      <td>30564.700000</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>188.317100</td>\n",
       "      <td>4.614840e-48</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.238099e-48</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>0.493243</td>\n",
       "      <td>0.506757</td>\n",
       "      <td>0.493243</td>\n",
       "      <td>-0.794589</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20784 rows  25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                          SUM_DEPTH  SUMSQ_DEPTH  MEAN_DEPTH     VAR_DEPTH  \\\n",
       "dDocent_Contig_109-16        305305    450193771  1027.96000  460647.00000   \n",
       "dDocent_Contig_156-11        328000    472002036  1104.38000  370832.00000   \n",
       "dDocent_Contig_160-14        236545    249594353   796.44800  206752.00000   \n",
       "dDocent_Contig_212-78        294044    383862802   990.04700  313329.00000   \n",
       "dDocent_Contig_220-41        158221    111715447   532.73100   92656.30000   \n",
       "...                             ...          ...         ...           ...   \n",
       "dDocent_Contig_396255-35        830         4432     3.50211       6.46292   \n",
       "dDocent_Contig_396263-18       1041         5737     3.81319       6.49806   \n",
       "dDocent_Contig_396286-6        5349       111427    18.01010      50.98300   \n",
       "dDocent_Contig_396287-2        1553        18101     6.21200      33.95090   \n",
       "dDocent_Contig_396299-19       2481        25667     8.38176      16.51480   \n",
       "\n",
       "                          N_DATA  N_GENOTYPE_FILTERED  N_MISS    F_MISS  \\\n",
       "dDocent_Contig_109-16        594                    0       0  0.000000   \n",
       "dDocent_Contig_156-11        594                    0       0  0.000000   \n",
       "dDocent_Contig_160-14        594                    0       0  0.000000   \n",
       "dDocent_Contig_212-78        594                    0       0  0.000000   \n",
       "dDocent_Contig_220-41        594                    0       0  0.000000   \n",
       "...                          ...                  ...     ...       ...   \n",
       "dDocent_Contig_396255-35     594                    0     120  0.202020   \n",
       "dDocent_Contig_396263-18     594                    0      48  0.080808   \n",
       "dDocent_Contig_396286-6      594                    0       0  0.000000   \n",
       "dDocent_Contig_396287-2      594                    0      94  0.158249   \n",
       "dDocent_Contig_396299-19     594                    0       2  0.003367   \n",
       "\n",
       "                                   QUAL N_ALLELES  ...   ChiSq_HWE  \\\n",
       "dDocent_Contig_109-16      37573.500000         2  ...   19.035410   \n",
       "dDocent_Contig_156-11     301506.000000         2  ...  223.728000   \n",
       "dDocent_Contig_160-14     313307.000000         2  ...  232.993500   \n",
       "dDocent_Contig_212-78       2530.050000         2  ...    0.360573   \n",
       "dDocent_Contig_220-41       3058.980000         2  ...    1.027682   \n",
       "...                                 ...       ...  ...         ...   \n",
       "dDocent_Contig_396255-35       0.047948         2  ...  140.154700   \n",
       "dDocent_Contig_396263-18    2370.400000         2  ...    0.411256   \n",
       "dDocent_Contig_396286-6     1768.980000         2  ...    8.282448   \n",
       "dDocent_Contig_396287-2      300.164000         2  ...    2.213322   \n",
       "dDocent_Contig_396299-19   30564.700000         2  ...  188.317100   \n",
       "\n",
       "                                 P_HWE P_HET_DEFICIT  P_HET_EXCESS A1_allele  \\\n",
       "dDocent_Contig_109-16     3.396307e-07  1.000000e+00  2.478008e-07         C   \n",
       "dDocent_Contig_156-11     2.161918e-62  1.000000e+00  2.027941e-62         T   \n",
       "dDocent_Contig_160-14     4.080681e-65  1.000000e+00  4.064197e-65         C   \n",
       "dDocent_Contig_212-78     1.000000e+00  1.000000e+00  7.183592e-01         G   \n",
       "dDocent_Contig_220-41     1.000000e+00  1.000000e+00  3.902908e-01         T   \n",
       "...                                ...           ...           ...       ...   \n",
       "dDocent_Contig_396255-35  8.440953e-11  8.440953e-11  1.000000e+00         A   \n",
       "dDocent_Contig_396263-18  7.021218e-01  7.867396e-01  3.441632e-01         G   \n",
       "dDocent_Contig_396286-6   1.436292e-03  1.000000e+00  8.767035e-04         A   \n",
       "dDocent_Contig_396287-2   2.308508e-01  1.000000e+00  1.385372e-01         A   \n",
       "dDocent_Contig_396299-19  4.614840e-48  1.000000e+00  3.238099e-48         G   \n",
       "\n",
       "                          A2_allele   A1_freq   A2_freq       MAF       Fis  \n",
       "dDocent_Contig_109-16             T  0.797980  0.202020  0.202020 -0.251055  \n",
       "dDocent_Contig_156-11             C  0.535354  0.464646  0.464646 -0.864780  \n",
       "dDocent_Contig_160-14             T  0.530303  0.469697  0.469697 -0.882540  \n",
       "dDocent_Contig_212-78             A  0.966330  0.033670  0.033670 -0.033101  \n",
       "dDocent_Contig_220-41             C  0.944444  0.055556  0.055556 -0.057041  \n",
       "...                             ...       ...       ...       ...       ...  \n",
       "dDocent_Contig_396255-35          T  0.962025  0.037975  0.037975  0.769493  \n",
       "dDocent_Contig_396263-18          A  0.802198  0.197802  0.197802 -0.036910  \n",
       "dDocent_Contig_396286-6           G  0.856902  0.143098  0.143098 -0.165029  \n",
       "dDocent_Contig_396287-2           C  0.914000  0.086000  0.086000 -0.091904  \n",
       "dDocent_Contig_396299-19          C  0.493243  0.506757  0.493243 -0.794589  \n",
       "\n",
       "[20784 rows x 25 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>N_ALLELES</th>\n",
       "      <th>N_CHR</th>\n",
       "      <th>A1_FREQ</th>\n",
       "      <th>A2_FREQ</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_109-16</th>\n",
       "      <td>dDocent_Contig_109</td>\n",
       "      <td>16</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>C:0.79798</td>\n",
       "      <td>T:0.20202</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_156-11</th>\n",
       "      <td>dDocent_Contig_156</td>\n",
       "      <td>11</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>T:0.535354</td>\n",
       "      <td>C:0.464646</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_160-14</th>\n",
       "      <td>dDocent_Contig_160</td>\n",
       "      <td>14</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>C:0.530303</td>\n",
       "      <td>T:0.469697</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_212-78</th>\n",
       "      <td>dDocent_Contig_212</td>\n",
       "      <td>78</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>G:0.96633</td>\n",
       "      <td>A:0.03367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_220-41</th>\n",
       "      <td>dDocent_Contig_220</td>\n",
       "      <td>41</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>T:0.944444</td>\n",
       "      <td>C:0.0555556</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396255-35</th>\n",
       "      <td>dDocent_Contig_396255</td>\n",
       "      <td>35</td>\n",
       "      <td>2</td>\n",
       "      <td>474</td>\n",
       "      <td>A:0.962025</td>\n",
       "      <td>T:0.0379747</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396263-18</th>\n",
       "      <td>dDocent_Contig_396263</td>\n",
       "      <td>18</td>\n",
       "      <td>2</td>\n",
       "      <td>546</td>\n",
       "      <td>G:0.802198</td>\n",
       "      <td>A:0.197802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396286-6</th>\n",
       "      <td>dDocent_Contig_396286</td>\n",
       "      <td>6</td>\n",
       "      <td>2</td>\n",
       "      <td>594</td>\n",
       "      <td>A:0.856902</td>\n",
       "      <td>G:0.143098</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396287-2</th>\n",
       "      <td>dDocent_Contig_396287</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>500</td>\n",
       "      <td>A:0.914</td>\n",
       "      <td>C:0.086</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396299-19</th>\n",
       "      <td>dDocent_Contig_396299</td>\n",
       "      <td>19</td>\n",
       "      <td>2</td>\n",
       "      <td>592</td>\n",
       "      <td>G:0.493243</td>\n",
       "      <td>C:0.506757</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20784 rows  6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM POS N_ALLELES N_CHR  \\\n",
       "dDocent_Contig_109-16        dDocent_Contig_109  16         2   594   \n",
       "dDocent_Contig_156-11        dDocent_Contig_156  11         2   594   \n",
       "dDocent_Contig_160-14        dDocent_Contig_160  14         2   594   \n",
       "dDocent_Contig_212-78        dDocent_Contig_212  78         2   594   \n",
       "dDocent_Contig_220-41        dDocent_Contig_220  41         2   594   \n",
       "...                                         ...  ..       ...   ...   \n",
       "dDocent_Contig_396255-35  dDocent_Contig_396255  35         2   474   \n",
       "dDocent_Contig_396263-18  dDocent_Contig_396263  18         2   546   \n",
       "dDocent_Contig_396286-6   dDocent_Contig_396286   6         2   594   \n",
       "dDocent_Contig_396287-2   dDocent_Contig_396287   2         2   500   \n",
       "dDocent_Contig_396299-19  dDocent_Contig_396299  19         2   592   \n",
       "\n",
       "                             A1_FREQ      A2_FREQ  \n",
       "dDocent_Contig_109-16      C:0.79798    T:0.20202  \n",
       "dDocent_Contig_156-11     T:0.535354   C:0.464646  \n",
       "dDocent_Contig_160-14     C:0.530303   T:0.469697  \n",
       "dDocent_Contig_212-78      G:0.96633    A:0.03367  \n",
       "dDocent_Contig_220-41     T:0.944444  C:0.0555556  \n",
       "...                              ...          ...  \n",
       "dDocent_Contig_396255-35  A:0.962025  T:0.0379747  \n",
       "dDocent_Contig_396263-18  G:0.802198   A:0.197802  \n",
       "dDocent_Contig_396286-6   A:0.856902   G:0.143098  \n",
       "dDocent_Contig_396287-2      A:0.914      C:0.086  \n",
       "dDocent_Contig_396299-19  G:0.493243   C:0.506757  \n",
       "\n",
       "[20784 rows x 6 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "frq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>OBS(HOM1/HET/HOM2)</th>\n",
       "      <th>E(HOM1/HET/HOM2)</th>\n",
       "      <th>ChiSq_HWE</th>\n",
       "      <th>P_HWE</th>\n",
       "      <th>P_HET_DEFICIT</th>\n",
       "      <th>P_HET_EXCESS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_109-16</th>\n",
       "      <td>dDocent_Contig_109</td>\n",
       "      <td>16</td>\n",
       "      <td>177/120/0</td>\n",
       "      <td>189.12/95.76/12.12</td>\n",
       "      <td>19.035410</td>\n",
       "      <td>3.396307e-07</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.478008e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_156-11</th>\n",
       "      <td>dDocent_Contig_156</td>\n",
       "      <td>11</td>\n",
       "      <td>21/276/0</td>\n",
       "      <td>85.12/147.76/64.12</td>\n",
       "      <td>223.728000</td>\n",
       "      <td>2.161918e-62</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>2.027941e-62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_160-14</th>\n",
       "      <td>dDocent_Contig_160</td>\n",
       "      <td>14</td>\n",
       "      <td>18/279/0</td>\n",
       "      <td>83.52/147.95/65.52</td>\n",
       "      <td>232.993500</td>\n",
       "      <td>4.080681e-65</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>4.064197e-65</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_212-78</th>\n",
       "      <td>dDocent_Contig_212</td>\n",
       "      <td>78</td>\n",
       "      <td>277/20/0</td>\n",
       "      <td>277.34/19.33/0.34</td>\n",
       "      <td>0.360573</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>7.183592e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_220-41</th>\n",
       "      <td>dDocent_Contig_220</td>\n",
       "      <td>41</td>\n",
       "      <td>264/33/0</td>\n",
       "      <td>264.92/31.17/0.92</td>\n",
       "      <td>1.027682</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.902908e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396255-35</th>\n",
       "      <td>dDocent_Contig_396255</td>\n",
       "      <td>35</td>\n",
       "      <td>226/4/7</td>\n",
       "      <td>219.34/17.32/0.34</td>\n",
       "      <td>140.154700</td>\n",
       "      <td>8.440953e-11</td>\n",
       "      <td>8.440953e-11</td>\n",
       "      <td>1.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396263-18</th>\n",
       "      <td>dDocent_Contig_396263</td>\n",
       "      <td>18</td>\n",
       "      <td>174/90/9</td>\n",
       "      <td>175.68/86.64/10.68</td>\n",
       "      <td>0.411256</td>\n",
       "      <td>7.021218e-01</td>\n",
       "      <td>7.867396e-01</td>\n",
       "      <td>3.441632e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396286-6</th>\n",
       "      <td>dDocent_Contig_396286</td>\n",
       "      <td>6</td>\n",
       "      <td>212/85/0</td>\n",
       "      <td>218.08/72.84/6.08</td>\n",
       "      <td>8.282448</td>\n",
       "      <td>1.436292e-03</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>8.767035e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396287-2</th>\n",
       "      <td>dDocent_Contig_396287</td>\n",
       "      <td>2</td>\n",
       "      <td>207/43/0</td>\n",
       "      <td>208.85/39.30/1.85</td>\n",
       "      <td>2.213322</td>\n",
       "      <td>2.308508e-01</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>1.385372e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>dDocent_Contig_396299-19</th>\n",
       "      <td>dDocent_Contig_396299</td>\n",
       "      <td>19</td>\n",
       "      <td>13/266/17</td>\n",
       "      <td>72.01/147.97/76.01</td>\n",
       "      <td>188.317100</td>\n",
       "      <td>4.614840e-48</td>\n",
       "      <td>1.000000e+00</td>\n",
       "      <td>3.238099e-48</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20784 rows  8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                          CHROM  POS OBS(HOM1/HET/HOM2)  \\\n",
       "dDocent_Contig_109-16        dDocent_Contig_109   16          177/120/0   \n",
       "dDocent_Contig_156-11        dDocent_Contig_156   11           21/276/0   \n",
       "dDocent_Contig_160-14        dDocent_Contig_160   14           18/279/0   \n",
       "dDocent_Contig_212-78        dDocent_Contig_212   78           277/20/0   \n",
       "dDocent_Contig_220-41        dDocent_Contig_220   41           264/33/0   \n",
       "...                                         ...  ...                ...   \n",
       "dDocent_Contig_396255-35  dDocent_Contig_396255   35            226/4/7   \n",
       "dDocent_Contig_396263-18  dDocent_Contig_396263   18           174/90/9   \n",
       "dDocent_Contig_396286-6   dDocent_Contig_396286    6           212/85/0   \n",
       "dDocent_Contig_396287-2   dDocent_Contig_396287    2           207/43/0   \n",
       "dDocent_Contig_396299-19  dDocent_Contig_396299   19          13/266/17   \n",
       "\n",
       "                            E(HOM1/HET/HOM2)   ChiSq_HWE         P_HWE  \\\n",
       "dDocent_Contig_109-16     189.12/95.76/12.12   19.035410  3.396307e-07   \n",
       "dDocent_Contig_156-11     85.12/147.76/64.12  223.728000  2.161918e-62   \n",
       "dDocent_Contig_160-14     83.52/147.95/65.52  232.993500  4.080681e-65   \n",
       "dDocent_Contig_212-78      277.34/19.33/0.34    0.360573  1.000000e+00   \n",
       "dDocent_Contig_220-41      264.92/31.17/0.92    1.027682  1.000000e+00   \n",
       "...                                      ...         ...           ...   \n",
       "dDocent_Contig_396255-35   219.34/17.32/0.34  140.154700  8.440953e-11   \n",
       "dDocent_Contig_396263-18  175.68/86.64/10.68    0.411256  7.021218e-01   \n",
       "dDocent_Contig_396286-6    218.08/72.84/6.08    8.282448  1.436292e-03   \n",
       "dDocent_Contig_396287-2    208.85/39.30/1.85    2.213322  2.308508e-01   \n",
       "dDocent_Contig_396299-19  72.01/147.97/76.01  188.317100  4.614840e-48   \n",
       "\n",
       "                          P_HET_DEFICIT  P_HET_EXCESS  \n",
       "dDocent_Contig_109-16      1.000000e+00  2.478008e-07  \n",
       "dDocent_Contig_156-11      1.000000e+00  2.027941e-62  \n",
       "dDocent_Contig_160-14      1.000000e+00  4.064197e-65  \n",
       "dDocent_Contig_212-78      1.000000e+00  7.183592e-01  \n",
       "dDocent_Contig_220-41      1.000000e+00  3.902908e-01  \n",
       "...                                 ...           ...  \n",
       "dDocent_Contig_396255-35   8.440953e-11  1.000000e+00  \n",
       "dDocent_Contig_396263-18   7.867396e-01  3.441632e-01  \n",
       "dDocent_Contig_396286-6    1.000000e+00  8.767035e-04  \n",
       "dDocent_Contig_396287-2    1.000000e+00  1.385372e-01  \n",
       "dDocent_Contig_396299-19   1.000000e+00  3.238099e-48  \n",
       "\n",
       "[20784 rows x 8 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hardy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "loci_out = os.path.join(analysis_dir,'loci_out.csv')\n",
    "frq_out = os.path.join(analysis_dir,'frq_out.csv')\n",
    "hardy_out = os.path.join(analysis_dir,'hardy_out.csv')\n",
    "\n",
    "loci_df.to_csv(path_or_buf=loci_out)\n",
    "frq_df.to_csv(path_or_buf=frq_out)\n",
    "hardy.to_csv(path_or_buf=hardy_out)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count     20784.000000\n",
       "mean       4612.489030\n",
       "std       13649.576592\n",
       "min         498.000000\n",
       "25%        1076.000000\n",
       "50%        1904.000000\n",
       "75%        3671.250000\n",
       "max      342977.000000\n",
       "Name: SUM_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    2.078400e+04\n",
       "mean     6.984903e+03\n",
       "std      3.922700e+04\n",
       "min      0.000000e+00\n",
       "25%      2.114525e+00\n",
       "50%      6.256240e+02\n",
       "75%      4.369460e+03\n",
       "max      2.459980e+06\n",
       "Name: QUAL, dtype: float64"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_df.QUAL.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAyMAAAIJCAYAAAC2k+IhAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nOzde7wkZX3v+88XEBTQ4bJNvKB7hIASrwGJCltAOBqMbDE6RNgREbwfRUU0UUQdFD1JIHgBlQSVSSTZg2LEAwLG7XB1VBRE9IjcR0VRQWCG4WaA3/mjqrXT9FqzZk3P1HSvz/v16letfur51VO9msWLL1VPPakqJEmSJGld26DrE5AkSZI0NxlGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJndio6xPQ2pPkBuARwLKOT0WSJEmTaz6woqqesLqFhpHJ9oiHPexhW+24445bdX0ikiRJmkxXXnkld99996xqDSOTbdmOO+641aWXXtr1eUiSJGlC7bzzzlx22WXLZlPrnBFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqxEZdn4Am0/x3faXrU1jnlv3ti7o+BUmSpLHilRFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqRNjFUaSbJ3kNUm+lOTaJHcnWZ7k4iSvTjL08yTZNcnZSW5ta65I8rYkG04z1r5Jzm+PvzLJt5McvIrzOzjJJW3/5W39vtP03zDJ4e353N2e39lJdp35b0WSJEkaT2MVRoD9gZOBZwHfBj4KfBF4CvBp4PNJ0l+QZD/gQmB34EvAicDGwEeAxcMGSfJm4Mz2uKe2Yz4GWJTkuClqjgMWAY9u+58KPBU4sz3eYP+04x/fns+J7fntDlzYnrckSZI0sTbq+gRW09XAi4GvVNUDvcYkRwKXAC8DXkoTUEjyCJpgcD+wZ1V9t21/L7AEWJDkgKpa3Hes+cBxwK3AM6tqWdv+AeA7wBFJvlhV3+yr2RU4ArgO2KWqbmvbjwUuBY5LclbvWK0DgAXAUmDvqrqnrTkJuBg4OcmSqrpjDX9nkiRJ0npprK6MVNWSqjqzP4i07b8ETmrf7tm3awHwSGBxL4i0/e8BjmrfvnFgmEOBTYAT+8NDGzA+3L59w0BN7/2HekGkrVkGfKI93iEDNb1xj+oFkbbmO8Bp7XkvQJIkSZpQYxVGVuE/2+19fW17tdtzh/S/ELgL2DXJJjOsOWegz6xqkjwU2LUd/6LVGEeSJEmaGON2m9ZQSTYCXtm+7Q8ET2y3Vw/WVNV9SW4AngxsC1w5g5qbktwJbJNk06q6K8lmwGOBlVV105DTu6bd7tDXth2wIXB9Vd334JKhNVNKcukUu540k3pJkiSpC5NyZeRvaSabn11VX+1rn9dul09R12vfYhY18wa2a2OMLabYL0mSJI29sb8ykuQtNJPHfwwc1PHpdKKqdh7W3l4x2Wkdn44kSZI0I2N9ZaR9ZO7HgB8Bz6uqWwe6DF7FGNRrv30WNcsHtmtjjNun2C9JkiSNvbENI0neBpwA/JAmiPxySLer2u2D5l6080yeQDPh/foZ1jwa2Ay4saruAqiqO4GfA5u3+wdt327756BcR/O44W3b85hJjSRJkjRRxjKMJPkbmkULL6cJIr+eouuSdrvPkH27A5sCS6vq3hnWvHCgz6xq2kf5Lm3Hf+5qjCNJkiRNjLELI+2ChX9Ls5jg3lV1yzTdTwduAQ5I8sy+YzwUOKZ9+6mBmlOAe4E3twsg9mq2BI5s3540UNN7/562X69mPvCm9ninDNT0xj2mPZ9ezS7Ay4GbaRdvlCRJkibRWE1gT3Iw8AGaW5wuAt6SZLDbsqpaBFBVK5K8liaUnJ9kMc3K6i+meYTv6TQLDP5OVd2Q5J3Ax4HvJjkN+C3NAoTbAP/Qv/p6W7M0yfHA24ErkpwObEwTKrYCDhtYfR1gMc1q8QuA7yU5E9i6rdkQeG1VrVjtX5IkSZI0JsYqjNDM8YDmP9bfNkWfC4BFvTdVdUaSPYD3AC8DHgpcSxMcPl5VNXiAqjohyTLgHTTrl2xAM0n+qKr652GDVtURSX5AcyXkdcADwGXAsVV11pD+leRAmtu1DgUOA+6hWYzxmKpaOvWvQZIkSRp/YxVGqmohsHAWdd8A/nw1a84EzlzNmkX0BaEZ9L+PZu7LR1ZnHEmSJGkSjN2cEUmSJEmTwTAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0YuzCSZEGSE5JclGRFkkpy6hR9F7X7p3t9faDmVavo/4YpxnpYkqOTXJXkniS/TvL5JDtO81m2SvLRJMuS3JvkF0k+m2SbNfstSZIkSeu/jbo+gVk4Cng6sBK4EXjSNH3PAJZNse8gYFvgnCn2fxm4fEj7dwcbkmwCfA3Yrd3/MeBxwP7Ai5LsVVXfHqjZGlgK7AAsARa3n+WQtuY5VXX9NJ9NkiRJGmvjGEYOpwkh1wJ7AOdN1bGqzqAJJP9Fki2AvwZ+CyyaovyMqppq36C30wSR04GXV9UD7TinteN/NslTe+2tD9MEkeOr6oi+c3sLTZj5JLDPDMeXJEmSxs7Y3aZVVedV1TVVVWtwmIOAhwH/XlW3rMn5JAnQu3Xrr/sDR1V9GbgI+GOa4NSr2bw9hzuBhQOHPBH4CfBnSbZdk3OTJEmS1mdjF0ZG5LXt9p+m6fOMJG9L8q4kB00zj2M74PHA1VV1w5D9vdvA9uprezZNGPpGVd3R37kNM19t3z5vug8hSZIkjbNxvE1rjSR5DvBUmvAw5S1ewFsH3t+f5NPA26rqnr72J7bbq6c4zjXtdoc1rJlSkkun2DXdfBpJkiSpU3Pxysjr2u3JU+y/ATiMJjBsBjwG+EuaifCvBz470H9eu10+xfF67VusYY0kSZI0UebUlZEk82iCxZQT16vqAuCCvqa7gC8k+RbwfeDAJH9XVd9fy6c7Y1W187D29orJTuv4dCRJkqQZmWtXRl4BbMosJq5X1c+As9u3u/ft6l3FmMdwvfbb17BGkiRJmihzLYz0Jq7/4yzrb263m/W1XdVup5rfsX277Z8fMpsaSZIkaaLMmTCS5Fk0iyVeXVXnz/Iwz2q3/YsRXgf8FNghyROG1Lyw3S7pa/sWcDewW5KHD5znBsAL2rfTTbCXJEmSxtqcCSP8fuL6dI/zJckzh7RtkOTdwHOAW4Bze/va9U5Oat/+fRsmenX7Ac8FfkTfPJSqWgl8juYKy8KB4d4MzAe+6grskiRJmmRjN4E9yUuAl7RvH9Vun5NkUfvzLVX1joGaRwAvB+4F/nkVQ3wnyQ9pJqv/nGb+xm7AU2gms/9VVa0YqDke2BdYAHw7yddp1h7Zv605dGD1dYAjgT2Btyd5BnAJsCOwH/Br4E2rOE9JkiRprI1dGAGeARw80LZt+4Jm9fJ3DOz/K5qrEItnMHH9OOBPaRYp3Ap4gOY2rE8Axw+7WlFV9yZ5PvAu4EDgcGAFcAbw/qr60ZCa37RrnryfJlw9F/gNcArwvqq6cRXnKUmSJI21sQsjVbWQB9/atKqaTwGfmmHfd67+WUFV3QW8r33NtOZWmsUVBxdYlCRJkibeXJozIkmSJGk9YhiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE6MXRhJsiDJCUkuSrIiSSU5dYq+89v9U70WTzPOwUkuSbIyyfIk5yfZd5r+GyY5PMkVSe5OcmuSs5PsOk3Nw5IcneSqJPck+XWSzyfZcfV+K5IkSdL42ajrE5iFo4CnAyuBG4EnzaDm+8AZQ9p/OKxzkuOAI9rjnwxsDBwAnJnksKo6caB/gMXAAuAq4ERgK+DlwIVJXlZVXx6o2QT4GrAb8F3gY8DjgP2BFyXZq6q+PYPPJkmSJI2lcQwjh9OEhGuBPYDzZlBzeVUtnMnB2ysZRwDXAbtU1W1t+7HApcBxSc6qqmV9ZQfQBJGlwN5VdU9bcxJwMXBykiVVdUdfzdtpgsjpwMur6oG25jSa4PTZJE/ttUuSJEmTZuxu06qq86rqmqqqtTTEG9rth3pBpB13GfAJYBPgkIGaN7bbo3pBpK35DnAa8EiasAL87kpKb5y/7g8c7RWUi4A/pglbkiRJ0kQauzAyS49J8vokR7bbp03Td692e+6QfecM9CHJQ4FdgbtoQsQqa4DtgMcDV1fVDTOskSRJkibKON6mNRvPb1+/k+R84OCq+mlf22bAY4GVVXXTkONc02536GvbDtgQuL6q7pthzRPb7dVTnO+wmikluXSKXTOZTyNJkiR1YtKvjNwFfBDYGdiyffXmmewJfL0NID3z2u3yKY7Xa9+igxpJkiRpokz0lZGq+jXwvoHmC5O8gGZi+bOA19A8yWpsVdXOw9rbKyY7rePTkSRJkmZk0q+MDNXeTvXp9u3ufbt6VyTmMVyv/fYOaiRJkqSJMifDSOvmdvu727Sq6k7g58DmSR49pGb7dts/1+M64H5g2yTDrjQNq7mq3U41J2RYjSRJkjRR5nIYeXa7vX6gfUm73WdIzQsH+tA+yncpsCnw3JnU0ASYnwI7JHnCDGskSZKkiTLRYSTJTkke9BmT7E2zeCLAqQO7T2q370myZV/NfOBNwL3AKQM1n2q3x7SP+u3V7EKzCvvNwBd77e0aKb1x/r7/HJPsRxNqfgRcsMoPKUmSJI2psZvAnuQlwEvat49qt89Jsqj9+Zaqekf78/HA9kmW0qzaDvA0fr9+x3uramn/8atqaZLjaVZIvyLJ6cDGNKFiK+CwgdXXARYDL6VZ2PB7Sc4Etm5rNgReW1UrBmqOB/Zta76d5Os0a4/sT/MUsENdfV2SJEmTbOzCCPAM4OCBtm3bF8BPgF4Y+RzwF8AuNLc+PQT4FfB54MSqGrZIIVV1RJIf0FwJeR3wAHAZcGxVnTWkfyU5kOZ2rUOBw4B7gAuBYwYDT1tzb5LnA+8CDqS5UrMCOAN4f1X9aNW/CkmSJGl8jV0YqaqFwMIZ9v0M8JlZjrMIWLQa/e8DPtK+ZlpzF82jhwcfPyxJkiRNvImeMyJJkiRp/WUYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOjHSMJLkIaM8niRJkqTJNeorIz9P8ndJ/mjEx5UkSZI0YUYdRjYA3glcleRrSV6WZMMRjyFJkiRpAow6jDwGeAVwEbA38HngxiQfSjJ/xGNJkiRJGmMjDSNV9duq+req2hN4EvBRYCPg3cC1Sc5Osl8SJ85LkiRJc9xaCwVVdXVVHQE8lt9fLdkH+Hfgp0kWJnnM2hpfkiRJ0vptrV+hqKrfAl8BvgT8AgjN7VzvA25I8tEkm6zt85AkSZK0flmrYSTJs5OcQhNCPgJsBnwceAZwKHAVcBjN7VySJEmS5pCNRn3AJA8HDgJeDzyF5krI94BPAv9WVXe3Xa9I8jngXGAB8MZRn4skSZKk9ddIw0iSzwB/CWwK3At8DvhkVV0yrH9V3Z/kfGCvUZ6HJEmSpPXfqG/TOgT4JfDXwDZV9aqpgkif84EPzHSAJAuSnJDkoiQrklSSU6fou32Sv0myJMnPkvw2ya+SfDnJ86aoeVV7zKleb5ii7mFJjk5yVZJ7kvw6yeeT7DjNZ9mqnTOzLMm9SX6R5LNJtpnp70OSJEkaV6O+TWufqvqP1Smoqm8A31iNkqOApwMrgRtpHiE8lQ8CLwd+BJwN3Ao8EXgx8OIkb62qj09R+2Xg8iHt3x1saCfgfw3Yrd3/MeBxwP7Ai5LsVVXfHqjZGlgK7AAsARa3n+WQtuY5VXX9NJ9NkiRJGmsjDSOrG0Rm6XCaEHItsAdw3jR9zwX+rqq+19+YZA+a8HBski9U1U1Das+oqkUzPKe30wSR04GXV9UD7TinAWcAn03y1F5768M0QeT49hHIvXN7C02Y+STNo5AlSZKkiTTS27SS7N3eZjR0/ZAkj2n37znbMarqvKq6pqpqBn0XDQaRtv0CmtvDNgZ2ne25ACQJ0Lt166/7A0dVfZlmfZU/pglOvZrNaSb53wksHDjkicBPgD9Lsu2anJskSZK0Phv1nJHDgF2r6hfDdrbtz2n7de0/2+19U+x/RpK3JXlXkoOmmcexHfB44OqqumHI/nPabf8k/WcDDwO+UVV39Hduw8xX27dD57VIkiRJk2DUc0Z2Av7PKvpcDLxgxOOuliT/HdgbuAu4cIpubx14f3+STwNvq6p7+tqf2G6vnuI417TbHdawZkpJLp1i13TzaSRJkqROjfrKyB/QLHA4nV+1/TrRTjb/V2ATYGFV3TbQ5QaaKzdPpFmk8TE0jyteRrN2ymcH+s9rt8unGLLXvsUa1kiSJEkTZdRXRpbTPEVqOo+jmSuxziXZkGbtk92A04DjBvu080ku6Gu6C/hCkm8B3wcOTPJ3VfX9dXDKM1JVOw9rb6+Y7LSOT0eSJEmakVFfGbkEeEmSRw3b2U5sf0nbb51qg8ipNI/b/TzwiplMgu+pqp/RPB4YYPe+Xb2rGPMYrtd++xrWSJIkSRNl1GHkBODhwEVJXtzeEkWSTZLsRzM/Y3NgqrU91ookDwH+N3AA8G/A/6qqqSauT+fmdrtZX9tV7Xaq+R3bt9v++SGzqZEkSZImykjDSLvOyAdpnjD1JeDOJDfT3Jb178C2wDFVde4ox51Oko2BL9BcEfkX4KCqun+Wh3tWu+1fjPA64KfADkmeMKTmhe12SV/bt4C7gd2SPHzgfDfg9xP8p1tDRZIkSRpro74yQlW9n2axvt6K5/Pa7VeAP2v3rxPtlZkvAfsBnwEOGVh4cFjNM4e0bZDk3TSPJb6FZjFFANpbvU5q3/59GyZ6dfsBz6VZAf6CvpqVNHNXNuPB64y8GZgPfNUV2CVJkjTJRj2BHfjdFZK1shp7kpfQzDsB6M1NeU6SRe3Pt1TVO9qfTwL+nCZA/Bx4X7NG4X9xflWd3/f+O0l+SDNZ/ec0YWo34Ck0k9n/qqpWDBzjeGBfYAHw7SRfp1l7ZP+25tAhIehIYE/g7UmeQTOPZkea4PRr4E2r+l1IkiRJ42ythJG17BnAwQNt27YvaFYv74WR3m1T/w143zTHPL/v5+OAP6VZpHAr4AGa27A+ARw/7GpFVd2b5PnAu4ADgcOBFcAZwPur6kdDan6T5DnA+2nC1XOB3wCnAO+rqhunOV9JkiRp7K21MJJkU2BLYMNh+6vqp7M5blUt5MG3Nk3Vd89ZHP+dq1vT1t1FE3imCz2DNbfSLK44uMCiJEmSNPFGHkaSHAT8Dc0tR1OptTG2JEmSpPEx0kCQ5FU0K5TfD1wE/AyYzSN0JUmSJE24UV+deAdwG/A/qurKER9bkiRJ0gQZ9aN9/wj4gkFEkiRJ0qqMOozcCtw74mNKkiRJmkCjDiNnAXtmyGIekiRJktRv1GHk3cAmwElJNh/xsSVJkiRNkFFPYP8CzYrjrwH+V5JrgNuH9Kuq2nvEY0uSJEkaI6MOI3v2/bwZzWrpw9SIx5UkSZI0ZkYaRqpq1Ld9SZIkSZpQhgdJkiRJnTCMSJIkSerEyMNIkg2SHJbkW0mWJ7mvb9+fJPlkkh1GPa4kSZKk8TLSMJJkY+BrwEeB7YA7gP41R24ADgX+apTjSpIkSRo/o74y8k7gecDRwB8Cn+7fWVW3AxcCfzbicSVJkiSNmVGHkb8CvlFVH6iqBxj+CN8bgMePeFxJkiRJY2bUYeQJwLdW0edWYKsRjytJkiRpzIw6jNwDbLGKPo9n+KrskiRJkuaQUYeRy4EXtBPZHyTJPJr5IpeMeFxJkiRJY2bUYeSfgMcB/5rkEf07kmwBLAK2BE4a8biSJEmSxsxGozxYVf3vJM8HXgW8GLgNIMl3gScDmwCfqKqzRzmuJEmSpPEz8kUPq+pQmrVEfgQ8kmadkZ2Aa4FXV9Vhox5TkiRJ0vgZ6ZWRnqpaBCxK8jCa27KWV9Wda2MsSZIkSeNprYSRnqq6G7h7bY4hSZIkaTyN/DYtSZIkSZqJkV4ZSXL9DLtWVW03yrElSZIkjZdR36a1AVBD2rcA5rU//wL4zxGPK0mSJGnMjPrRvvOn2pfkj4CPA5vRLHwoSZIkaQ5bZ3NGqupa4KXAY4H3r6txJUmSJK2f1ukE9qq6B/gacOC6HFeSJEnS+qeLp2ndBzyqg3ElSZIkrUfWaRhJ8t+AvwB+ti7HlSRJkrT+GfWjfd83zTiPA/ajearWu0c5riRJkqTxM+pH+y5cxf4VwDFV9fcjHleSJEnSmBl1GHneFO0PALcBP66q+0Y8piRJkqQxNNI5I1V1wRSvi6rqh6MIIkkWJDkhyUVJViSpJKeuombXJGcnuTXJ3UmuSPK2JBtOU7NvkvOTLE+yMsm3kxy8inEOTnJJ2395W7/vNP03THJ4ez53t+d3dpJdV/2bkCRJksZbF0/TWlNHAW8GngH8fFWdk+wHXAjsDnwJOBHYGPgIsHiKmjcDZwJPAU4FTgYeAyxKctwUNccBi4BHt/1PBZ4KnNkeb7B/2vGPb8/nxPb8dgcubM9bkiRJmlijnsD++NnWVtVPZ9j1cOBG4FpgD+C8ac7nETTB4H5gz6r6btv+XmAJsCDJAVW1uK9mPnAccCvwzKpa1rZ/APgOcESSL1bVN/tqdgWOAK4Ddqmq29r2Y4FLgeOSnNU7VusAYAGwFNi7XYOFJCcBFwMnJ1lSVXfM8PciSZIkjZVRXxlZBtwwi9f1Mx2gqs6rqmuqqmbQfQHwSGBxL4i0x7iH5goLwBsHag4FNgFO7A8PbcD4cPv2DQM1vfcf6gWRtmYZ8In2eIcM1PTGPaoXRNqa7wCntee9YJWfUJIkSRpTo57A/i/AfJpbjZYDlwO/pFnk8Bk0j/W9gCa0rAt7tdtzh+y7ELgL2DXJJlV17wxqzhnoM5NxzgHe2/Z5P0CShwK7tuNfNEXNQW3NKUP2S5IkSWNv1GHk/wG+STMf4+iqWtHb0d4ydTTwSuD1VXX1iMce5ont9kFjVdV9SW4AngxsC1w5g5qbktwJbJNk06q6K8lmwGOBlVV105BzuKbd7tDXth2wIXD9FJP6h9VMKcmlU+x60kzqJUmSpC6M+jatvwV+UFVH9AcRgKpaUVWHA/9f229dmNdul0+xv9e+xSxq5g1s18YYW0yxX5IkSRp7o74ysjtw0ir6XAy8fsTjzmlVtfOw9vaKyU7r+HQkSZKkGRn1lZFNaOaHTOfRbb91YfAqxqBe++2zqFk+sF0bY9w+xX5JkiRp7I06jHwPOCDJnwzbmWRn4OXAZSMedypXtdsHzb1IshHwBOA+/uvTvKareTSwGXBjVd0FUFV30qx3snm7f9D27bZ/Dsp1NI8b3rY9j5nUSJIkSRNl1GHkaJqrHt9K8tkkr0rywnZ7Cs2aGg9p+60LS9rtPkP27Q5sCizte5LWqmpeONBnVjXto3yXtuM/dzXGkSRJkibGSMNIVf0fmsX8VgKvAj4DnNVuD27bD6iqr49y3GmcDtxCc7Xmmb3G9tG6x7RvPzVQcwpwL/DmdgHEXs2WwJHt28F5Mb3372n79WrmA29qjzf4iN7euMe059Or2YXm6tHNwBdX8fkkSZKksTXqCexU1elJzgH2o5k8PY9mjsRlwJfb25pmLclLgJe0b3vzU56TZFH78y1V9Y72XFYkeS1NKDk/yWKaldVfTPMI39NpFhjsP/8bkrwT+Djw3SSnAb+lWYBwG+Af+ldfb2uWJjkeeDtwRZLTgY1pQsVWwGEDq68DLAZe2h73e0nOBLZuazYEXjv4RDJJkiRpkow8jMDv5lH8W/satWfQXGXpt237AvgJ8I6+czkjyR7Ae4CXAQ8FrqUJDh8ftpJ7VZ2QZFl7nFfSXEH6Ec1q6f887KSq6ogkP6C5EvI64AGaAHZsVZ01pH8lOZDmdq1DgcOAe2gWYzymqpau+lchSZIkja+1EkZ62luWNq+qn43qmFW1EFi4mjXfAP58NWvOBM5czZpFwKLV6H8fzQKRH1mdcSRJkqRJMOoJ7CTZPMk/JPklzXyNG/r2PSvJ2Ulc+0KSJEma40YaRpLMA74JHA78ArgSSF+XH9A8PeqriEMAABxtSURBVOrAUY4rSZIkafyM+srIe4AnA6+qqp2AL/TvbNfmuADYe8TjSpIkSRozow4jLwW+WlX/Mk2fnwCPHfG4kiRJksbMqMPINsAVq+izkuZxv5IkSZLmsFGHkTuAP1hFnyfQTGyXJEmSNIeNOox8B9g3ycOH7UzyaJpH7F484nElSZIkjZlRh5GP0awifnaSHft3tO+/QLPo4MdHPK4kSZKkMTPSRQ+r6qtJjgbeD/wQ+E+AJLcAW9I85vdvXF1ckiRJ0sgXPayqo2ke3fv/ArcB9wMFnA38X1V17KjHlCRJkjR+RnplJMnuwIqqOg84b5THliRJkjRZRn1l5DzgdSM+piRJkqQJNOowcgtw94iPKUmSJGkCjTqMnA/sOuJjSpIkSZpAow4jRwFPTPLBJA8Z8bElSZIkTZCRTmAH3k3zSN8jgVcn+T7wS5qnafWrqnr1iMeWJEmSNEZGHUZe1ffzo9rXMAUYRiRJkqQ5bNRh5AkjPp4kSZKkCbXGYSTJK4HLq+qKqvrJCM5JkiRJ0hwwignsi4CX9DckOTjJkhEcW5IkSdKEGvXTtHrmA3uspWNLkiRJmgBrK4xIkiRJ0rQMI5IkSZI6YRiRJEmS1IlRhZHBRQ0lSZIkaVqjWmdkYZKFg41J7p+if1XVqNc4kSRJkjRGRhUIspb7S5IkSZowaxxGqsp5J5IkSZJWm0FCkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnZj4MJLkVUlqFa/7+/rPX0XfxdOMdXCSS5KsTLI8yflJ9p2m/4ZJDk9yRZK7k9ya5Owku4769yBJkiStb0a1Avv67HLg6Cn2PRfYCzhnyL7vA2cMaf/hsAMlOQ44ArgROBnYGDgAODPJYVV14kD/AIuBBcBVwInAVsDLgQuTvKyqvjz9R5MkSZLG18SHkaq6nCaQPEiSb7Y//tOQ3ZdX1cKZjNFeyTgCuA7Ypapua9uPBS4FjktyVlUt6ys7gCaILAX2rqp72pqTgIuBk5Msqao7ZnIOkiRJ0riZ+Nu0ppLkqcCzgZ8DX1nDw72h3X6oF0QA2vDxCWAT4JCBmje226N6QaSt+Q5wGvBImrAiSZIkTaQ5G0aA17Xbz1TV/UP2PybJ65Mc2W6fNs2x9mq35w7Zd85AH5I8FNgVuAu4aCY1kiRJ0qSZ+Nu0hknyMOAVwP3Ap6fo9vz21V93PnBwVf20r20z4LHAyqq6achxrmm3O/S1bQdsCFxfVffNsGZKSS6dYteTZlIvSZIkdWGuXhn5S2AL4Nyq+tnAvruADwI7A1u2rz2A84A9ga+3AaRnXrtdPsVYvfYt1rBGkiRJmihz8soIv79F6x8Hd1TVr4H3DTRfmOQFNBPLnwW8BvjYWj3D1VBVOw9rb6+Y7LSOT0eSJEmakTl3ZSTJk2nma9wInD3TuvZ2qt4tXbv37epdxZjHcL3229ewRpIkSZoocy6MsOqJ69O5ud3+7jatqrqT5olcmyd59JCa7dvt1X1t19HMV9k2ybCrU8NqJEmSpIkyp8JI+xSrg2iCwGdmcYhnt9vrB9qXtNt9htS8cKAP7aN8lwKb0iy8uMoaSZIkadLMqTAC7E8zIf2cIRPXAUiyU5IH/V6S7A0c3r49dWD3Se32PUm27KuZD7wJuBc4ZaDmU+32mDYk9Wp2oVmF/Wbgi6v+SJIkSdJ4mmsT2Hu3aA1bcb3neGD7JEtp5pUAPI3fr/nx3qpa2l9QVUuTHA+8HbgiyenAxjShYivgsIHV1wEWAy+lWdjwe0nOBLZuazYEXltVK1b/I0qSJEnjYc6EkSQ7Av+DVU9c/xzwF8AuNLdLPQT4FfB54MSqGrZIIVV1RJIf0FwJeR3wAHAZcGxVnTWkfyU5kOZ2rUOBw4B7gAuBYwYDjyRJkjRp5kwYqaorgcyg32eY3XwSqmoRsGg1+t8HfKR9SZIkSXPKXJszIkmSJGk9YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE7MiTCSZFmSmuL1yylqdk1ydpJbk9yd5Iokb0uy4TTj7Jvk/CTLk6xM8u0kB6/i3A5Ocknbf3lbv++afmZJkiRpfbdR1yewDi0HPjqkfeVgQ5L9gC8C9wCnAbcC/xP4CLAbsP+QmjcDJwC/AU4FfgssABYleWpVvWNIzXHAEcCNwMnAxsABwJlJDquqE1f/Y0qSJEnjYS6FkdurauGqOiV5BE0wuB/Ys6q+27a/F1gCLEhyQFUt7quZDxxHE1qeWVXL2vYPAN8Bjkjyxar6Zl/NrjRB5Dpgl6q6rW0/FrgUOC7JWb1jSZIkSZNmTtymtZoWAI8EFveCCEBV3QMc1b5940DNocAmwIn94aENGB9u375hoKb3/kO9INLWLAM+0R7vkDX5IJIkSdL6bC6FkU2SvCLJkUnemuR5U8z/2Kvdnjtk34XAXcCuSTaZYc05A33WpEaSJEmaGHPpNq1HAZ8baLshySFVdUFf2xPb7dWDB6iq+5LcADwZ2Ba4cgY1NyW5E9gmyaZVdVeSzYDHAiur6qYh53pNu91hJh8syaVT7HrSTOolSZKkLsyVKyOnAHvTBJLNgKcC/wjMB85J8vS+vvPa7fIpjtVr32IWNfMGtqszhiRJkjRR5sSVkao6eqDph8AbkqykmUS+EPiLdX1eo1JVOw9rb6+Y7LSOT0eSJEmakblyZWQqJ7Xb3fvaBq9iDOq13z6LmuUD29UZQ5IkSZoocz2M3NxuN+tru6rdPmi+RpKNgCcA9wHXz7Dm0e3xb6yquwCq6k7g58Dm7f5B27fbB81BkSRJkibFXA8jz263/cFiSbvdZ0j/3YFNgaVVde8Ma1440GdNaiRJkqSJMfFhJMmO7dOrBtvnA70Vzk/t23U6cAtwQJJn9vV/KHBM+/ZTA4c7BbgXeHN73F7NlsCR7duTBmp679/T9us/rze1xztlmo8mSZIkjbW5MIH95TQroF8I/AS4A9gOeBHwUOBsmtXTAaiqFUleSxNKzk+ymGZl9RfTPML3dOC0/gGq6oYk7wQ+Dnw3yWnAb2kWUNwG+If+1dfbmqVJjgfeDlyR5HRg4/Z8twIOc/V1SZIkTbK5EEbOowkRfwLsRjN/43bgYpp1Rz5XVdVfUFVnJNkDeA/wMprQci1NcPj4YP+25oQky4B3AK+kuer0I+CoqvrnYSdWVUck+QHNlZDXAQ8AlwHHVtVZa/i5JUmSpPXaxIeRdkHDC1bZ8cF13wD+fDVrzgTOXM2aRcCi1amRJEmSJsHEzxmRJEmStH4yjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJyY+jCTZOslrknwpybVJ7k6yPMnFSV6dZIOB/vOT1DSvxdOMdXCSS5KsbMc4P8m+0/TfMMnhSa5oz+vWJGcn2XWUvwNJkiRpfbRR1yewDuwPfAq4CTgP+Cnwh8BLgU8DL0yyf1XVQN33gTOGHO+HwwZJchxwBHAjcDKwMXAAcGaSw6rqxIH+ARYDC4CrgBOBrYCXAxcmeVlVfXn1P64kSZI0HuZCGLkaeDHwlap6oNeY5EjgEuBlNMHkiwN1l1fVwpkM0F7JOAK4Dtilqm5r248FLgWOS3JWVS3rKzuAJogsBfauqnvampOAi4GTkyypqjtW7+NKkiRJ42Hib9OqqiVVdWZ/EGnbfwmc1L7dcw2HeUO7/VAviLRjLAM+AWwCHDJQ88Z2e1QviLQ13wFOAx5JE1YkSZKkiTTxYWQV/rPd3jdk32OSvD7Jke32adMcZ692e+6QfecM9CHJQ4FdgbuAi2ZSI0mSJE2auXCb1lBJNgJe2b4dFiKe3776a84HDq6qn/a1bQY8FlhZVTcNOc417XaHvrbtgA2B66tqWBAaVjOlJJdOsetJM6mXJEmSujCXr4z8LfAU4Oyq+mpf+13AB4GdgS3b1x40k9/3BL7eBpCeee12+RTj9Nq3WMMaSZIkaaLMySsjSd5CM+H8x8BB/fuq6tfA+wZKLkzyApqJ5c8CXgN8bB2c6oxU1c7D2tsrJjut49ORJEmSZmTOXRlJ8maaIPEj4HlVdetM6trbqT7dvt29b1fvKsY8huu1376GNZIkSdJEmVNhJMnbgBNo1gp5XvtErdVxc7v93W1aVXUn8HNg8ySPHlKzfbu9uq/tOuB+YNt27spMaiRJkqSJMmfCSJK/AT4CXE4TRH49i8M8u91eP9C+pN3uM6TmhQN9aB/luxTYFHjuTGokSZKkSTMnwkiS99JMWL+UZoHBW6bpu1OSB/1ekuwNHN6+PXVgd2+9kvck2bKvZj7wJuBe4JSBmk+122PaR/32anahWYX9Zh68EKMkSZI0MSZ+AnuSg4EP0NwWdRHwliSD3ZZV1aL25+OB7ZMsBW5s257G79f8eG9VLe0vrqqlSY4H3g5ckeR0YGOaULEVcNjA6usAi2lWfl8AfC/JmcDWbc2GwGurasVsP7ckSZK0vpv4MAI8od1uCLxtij4XAIvanz8H/AWwC83tUg8BfgV8HjixqoYtUkhVHZHkBzRXQl4HPABcBhxbVWcN6V9JDqS5XetQ4DDgHuBC4JjBwCNJkiRNmokPI1W1EFi4Gv0/A3xmlmMt4vehZib976OZx/KR2YwnSZIkjbM5MWdEkiRJ0vrHMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjAiSZIkqROGEUmSJEmdMIxIkiRJ6oRhRJIkSVInDCOSJEmSOmEYkSRJktQJw4gkSZKkThhGJEmSJHXCMCJJkiSpE4YRSZIkSZ0wjEiSJEnqhGFEkiRJUicMI5IkSZI6YRiRJEmS1AnDiCRJkqROGEYkSZIkdcIwIkmSJKkThhFJkiRJnTCMdCzJNkk+m+QXSe5NsizJR5Ns2fW5SZIkSWvTRl2fwFyWZDtgKfAHwJeBHwN/CrwV2CfJblX1mw5PUZIkSVprvDLSrU/SBJG3VNVLqupdVbUX8BHgicCHOj07SZIkaS0yjHSkvSryAmAZ8ImB3e8H7gQOSrLZOj41SZIkaZ0wjHTnee32P6rqgf4dVXUH8A1gU+DZ6/rEJEmSpHXBOSPdeWK7vXqK/dfQXDnZAfj6dAdKcukUu55+5ZVXsvPOO8/uDNfATT9fvs7H7Nomi97a9SloHXjKY+d1fQqSJK1XrrzySoD5s6k1jHSn9180U/1Xe699izUY4/677757+WWXXbZsDY4xG09qtz9ex+OqG3Pq+77sV12fQefm1Pc9x/ldzy1+33PLqL/v+cCK2RQaRiZAVa37Sx/T6F2pWd/OS2uH3/fc4vc9d/hdzy1+33PL+vR9O2ekO70rH1Pd89Frv30dnIskSZK0zhlGunNVu91hiv3bt9up5pRIkiRJY80w0p3z2u0LkvyX7yHJw4HdgLuAb63rE5MkSZLWBcNIR6rqOuA/aCb8vGlg99HAZsDnqurOdXxqkiRJ0jrhBPZu/d/AUuDjSfYGrgSeRbMGydXAezo8N0mSJGmtSlV1fQ5zWpLHAR8A9gG2Bm4CvgQcXVW3dXlukiRJ0tpkGJEkSZLUCeeMSJIkSeqEYUSSJElSJwwjkiRJkjphGJEkSZLUCcOIJEmSpE4YRiRJkiR1wjCiVUqyTZLPJvlFknuTLEvy0SRbruZxtmrrlrXH+UV73G3W1rlr9Y3i+05yfpKa5vXQtfkZNDNJFiQ5IclFSVa0382pszzWSP49obVnVN93+91O9bf9y7Vx7lo9SbZO8pokX0pybZK7kyxPcnGSVydZrf/+8+97/TXK77qrv21XYNe0kmxHs0r8HwBfBn4M/CnwVmCfJLtV1W9mcJyt2+PsACwBFgNPAg4BXpTkOVV1/dr5FJqpUX3ffY6eov2+NTpRjcpRwNOBlcCNNH+Tq20t/HOjtWMk33drOfDRIe0r1+CYGp39gU/RLKR8HvBT4A+BlwKfBl6YZP+awWJz/n2v90b2XbfW/d92VfnyNeUL+CpQwGED7ce37SfN8Dj/2Pb/h4H2t7Tt53b9WX2N9Ps+v/nXS/efyde039PzgO2BAHu23/GpszjOSP658TU23/cyYFnXn8fXtN/RXsD/BDYYaH8UzX+sFvCyGR7Lv+/1+DXi77qTv21XYNeU2v8bci3NP5zbVdUDffseTpPCA/xBVd05zXE2B34NPAA8uqru6Nu3AXA98N/bMbw60pFRfd9t//OBPaoqa+2ENVJJ9qT5v2r/WlWvWI26kf1zo3Vntt93W7sMoKrmj/zEtNYlORL4EHBiVR22ir7+fY+x1fmu2/7LYN3/bTtnRNN5Xrv9j/5/AQG0geL/b+9eQy0r6ziOf394dyKdCc0XVkckHWlMsZsoKCKOpUGjGV0s03ojaWklVGI0YVBQZGZBL0olhzQwyhdaKjWaWSZFgpbX0hHL0nEcbfKe/16stWW3Pfe9z1lnH78fWDysy372s3nWczj/tZ7LzcCuwKEz5HMosAtwc38g0ubzIs1Tl/7vUzdGVd8vSfL+JJ9P8pkk70qy0+iKqyVi5PeNxsJOST6c5NwkZyU5Ksl2XRdKs/J8m86mu6zte7zNpa57Fr1tO2ZE09m/Te+Z4vy9wFqacSC/HDIf2nzUnVHVd78rBvYfSXJGVV05j/JpaVqI+0ZL317AZQPH7k9yWlXd2EWBNLMk2wOntLu/mMVHbN9jah513bPobds3I5rObm36xBTne8d3X6R8tLBGWU9X0fRh3Zvmrdhq4KvtZ3+c5J1DlFNLi+37lecS4Giaf1pWAAfSjAucAH6e5KDuiqYZfA1YA1xTVdfOdDG273E217qGjtq2b0YkjVxVXTBw6G7g3CT/AC6iCUzm8qRG0hJRVYOz5N0BnJ5kG/BZYD1wwmKXS9NL8ima+rkL+EjHxdECmm9dd9W2fTOi6fSeeOw2xfne8a2LlI8W1mLU0/dp+q4e3A5+1Pizfavne216RKel0MskORO4EPgLcFRVbZnlR23fY2aIup7OgrZtgxFN5+42nWosxxvbdKq+pKPORwtrweupqp4BepMYrJhvPlpSbN/qebRNbdtLSJKzad5I30Hzz+lcFq+zfY+RIet6Ogvatg1GNJ2Nbbp2cAXP9qn24cBTwC0z5HML8DRw+ODT8DbftQPfp26Mqr6nlGR/YCVNQLJ5vvloSVnw+0ZjozejklO0LxFJPgdcANxG88/pI3PMwvY9JkZQ19NZ0LZtMKIpVdVfgetoBi6dMXD6yzQR8mX9c4snWZ3k/1b1raptNDMzrKDpb9jvzDb/a11jpFujqu8k+yRZNZh/kj1oBscBXFFVrsI+RpLs0Nb3vv3H53PfaOmbqr6THJDkZU9Hk0wA32l3Nyx8CTWTJF+kGcT8R+DoqpryAZDte7yNoq67bNsueqhptTfrb4E9aWZIuhN4B83c4/cAh1XVY33XF8DgYndJXtPmsx/wK+BW4ADgPTQLIh7W/tFTh0ZR30lOpelf+huapyhbgNcDx9H0L/4DcExV2ce4Y0nWAeva3b2AY2nq7Kb22OaqOqe9dgK4H9g0uCDWXO8bdWMU9Z1kPc1A1l8Dm2jecu4LHA/sDFwDnFBVzy3oj9G0knwUuBT4L023nclmw3qgqi5tr5/A9j2WRlXXnbbtxV7y3W38NuB1NE+0Hwaea2/SbwErJ7m2mttq0nxW0Qyq2tTm8zBwMbB317/RbXT1TTMV4KXA7cBjNIsubaH5h+eTwI5d/0a3l+pqfa8Op9ge6Lt2YvDYfO8bt/Gtb+BI4HKaWXq2tu37UeB6mjUN0vXvdJtVXRdww0z13Xfe9r1Et1HVdZdt2zcjkiRJkjrhmBFJkiRJnTAYkSRJktQJgxFJkiRJnTAYkSRJktQJgxFJkiRJnTAYkSRJktQJgxFJkiRJnTAYkSRJktQJgxFJkiRJnTAYkSRJktQJgxFJkiRpDCU5KclFSW5K8mSSSrJhAb7nkCQ/SvJQkmeT/CvJjUlOGTbv7UdRQEmSJEmL7jzgIGAb8BCwetRfkORM4ELgceBq4O/AKmANcBzww2HyNxiRJEmSxtOnaYKQ+4AjgY2jzDzJWuDbwPXASVX174HzOwz7HXbTkiRJksZQVW2sqnurqmb7mSQfTLIxydYkzyS5M8l5SXaa5PKvA08DHxoMRNrvf36I4gO+GZEkSZJeEZJcDJxG8zblJ8BW4FDgfODoJMdU1QvttWuANwM/A7YkOQp4C1DAbcDGqnpx2DIZjEiSJEnLXJJTaQKRnwInV9XTfefWA18CzqAZHwLwtjZ9BLgBOGIgy9uTnFhV9w1TLrtpSZIkScvfWcALwMf6A5HW+cBjwMl9x/Zs048DE8DxwG7AfsAG4EDg6iQ7DlMo34xIkiRJy1iSXWlm3doMnJ1kssueBQ7o2++9tNgO+EBV/a7df7Kd0nc18FbgvcDl8y2bwYgkSZK0vK0EAuxB0x1rNra26T/7AhEAqqqSXEUTjLydIYIRu2lJkiRJy9sTbfqnqsp0W99n7m7TrUzu8TbdZZiCGYxIkiRJy1hVbQP+DLwpyapZfuwW4D/ARJIVk5xf06b3D1M2gxFJkiRp+fsmsCNwcZLdB08mWZnkkN5+VT0F/ADYGfhK+gaaJDkQOJVmQPyVwxQqc1gjRZIkSdISkWQdsK7d3Qs4FvgbcFN7bHNVndN3/XeBTwBbgGuBB4FVwD40U/deUlWn913/auBG4GDg98DNwGuBE2m6Z51dVb2pgOf3GwxGJEmSpPHTtz7IVDZV1cTAZ94NnE4z8Hx3msDkQeA6YENV3TVw/auALwDvA95AsyL7rcA3quq6oX+DwYgkSZKkLjhmRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVInDEYkSZIkdcJgRJIkSVIn/gcTxcaIfPQjBwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 260,
       "width": 401
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.QUAL.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAzgAAAHwCAYAAABqqMpyAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAAWJQAAFiUBSVIk8AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+j8jraAAAgAElEQVR4nOzdebhlVX3u++9rIX2ki4kNegsIKLEJATkKHAHhajByxCgESFSEKOpVDFCYKKAWirkmEFRAJUGhEklOoXiECwLGK72l0onoAelLBVFEoGgKMMDv/DHn0sVird1Uraqdmvv7eZ79jL3GHL85xl4hPvU+c84xU1VIkiRJUhc8baYXIEmSJEnjYsCRJEmS1BkGHEmSJEmdYcCRJEmS1BkGHEmSJEmdYcCRJEmS1BkGHEmSJEmdYcCRJEmS1BkGHEmSJEmdYcCRJEmS1BkGHEmSJEmdYcCRJEmS1BmrzfQCtOIkuQ14BrB4hpciSZKk7poL3F9Vm8z0QsCA03XPWGuttTbccsstN5zphUiSJKmbrr/+eh5++OGZXsZvGHC6bfGWW2654VVXXTXT65AkSVJHbbPNNlx99dWLZ3odPT6DI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOsOAI0mSJKkzDDiSJEmSOmO1mV6AumnuB74200tY6RZ/4nUzvQRJkqRZzys4kiRJkjrDgCNJkiSpMww4kiRJkjpjlQo4STZK8vYkX01yc5KHkyxJclmSv0oy9O9Jsn2Sc5Pc09Zcm+TgJHMmmGv3JBe1538wyXeT7DfJ+vZLcnk7fklbv/sE4+ckOaRdz8Pt+s5Nsv3UvxVJkiRJPatUwAH2Ak4GXg58F/gU8BXgxcDngS8lSX9Bkj2AS4Adga8CJwKrA58EFg6bJMl7gbPb857WzvkcYEGSY0fUHAssAJ7djj8NeAlwdnu+wfFp5z+uXc+J7fp2BC5p1y1JkiRpGla1XdRuBF4PfK2qnuh1JjkcuBx4E/BGmtBDkmfQhI3HgZ2r6sq2/0PABcCeSfapqoV955oLHAvcA7ysqha3/R8FrgDmJflKVX27r2Z7YB5wC7BtVd3b9h8DXAUcm+Sc3rla+wB7AouAXavqkbbmJOAy4OQkF1TVA8v5nUmSJEmzxip1BaeqLqiqs/vDTdv/c+Ck9uPOfYf2BJ4JLOyFm3b8I8CR7cd3D0xzALAGcGJ/IGlDy9+1H981UNP7/PFeuGlrFgOfac+3/0BNb94je+GmrbkCOL1d955IkiRJmrJVKuBM4j/b9rG+vl3a9vwh4y8BlgLbJ1ljijXnDYxZppokawLbt/NfOo15JEmSJE1gVbtFbagkqwFvbT/2h4wXtO2NgzVV9ViS24AXAZsC10+h5s4kDwEbJ1m7qpYmWQd4LvBgVd05ZHk3te0WfX2bAXOAW6vqsaeWDK0ZKclVIw69cCr1kiRJUld05QrOJ2g2BDi3qr7e179e2y4ZUdfrX38ZatYbaFfEHOuPOC5JkiRpiFX+Ck6S99E84P8j4C0zvJwZUVXbDOtvr+xsvZKXI0mSJM2YVfoKTrv98qeB64BXVdU9A0MGr7YM6vXftww1SwbaFTHHfSOOS5IkSRpilQ04SQ4GTgB+SBNufj5k2A1t+5RnWdrndjah2ZTg1inWPBtYB7i9qpYCVNVDwB3Auu3xQZu3bf8zPbfQbF29abuOqdRIkiRJmsQqGXCS/C3NizqvoQk3d40YekHb7jbk2I7A2sCiqnp0ijWvHRizTDXtttCL2vlfOY15JEmSJE1glQs47Us6P0HzAs1dq+ruCYafAdwN7JPkZX3nWBM4uv34uYGaU4FHgfe2L/3s1WwAHN5+PGmgpvf5iHZcr2Yu8J72fKcO1PTmPbpdT69mW2Bv4Je0LyyVJEmSNDWr1CYDSfYDPkpze9elwPuSDA5bXFULAKrq/iTvoAk6FyVZCNwDvJ5mO+gzaF6q+RtVdVuS9wPHA1cmOR34Nc1LNzcG/rGqvj1QsyjJccChwLVJzgBWpwkqGwIH9b80tLUQeGN73u8lORvYqK2ZA7yjqu6f9pckSZIkzWKrVMCheWYGmgBw8IgxFwMLeh+q6swkOwFHAG8C1gRupgkjx1dVDZ6gqk5Ishg4jOb9Ok+j2cjgyKr6l2GTVtW8JD+guWJzIPAEcDVwTFWdM2R8JdmX5la1A4CDgEdoXkB6dFUtGv01SJIkSRpmlQo4VTUfmL8Mdd8C/nSaNWcDZ0+zZgF94WoK4x+jeZbok9OZR5IkSdJwq9wzOJIkSZI0igFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1xioXcJLsmeSEJJcmuT9JJTltxNgF7fGJfr45UPO2Sca/a8RcayU5KskNSR5JcleSLyXZcoK/ZcMkn0qyOMmjSX6W5JQkGy/ftyRJkiTNTqvN9AKWwZHAHwEPArcDL5xg7JnA4hHH3gJsCpw34vhZwDVD+q8c7EiyBvANYIf2+KeB5wF7Aa9LsktVfXegZiNgEbAFcAGwsP1b9m9rtquqWyf42yRJkiQNWBUDziE0weZmYCfgwlEDq+pMmpDzJEnWB/4G+DWwYET5mVU16tigQ2nCzRnA3lX1RDvP6e38pyR5Sa+/9Xc04ea4qprXt7b30QSkzwK7TXF+SZIkSayCt6hV1YVVdVNV1XKc5i3AWsD/qqq7l2c9SQL0blv7m/4QU1VnAZcCf0gTxno167ZreAiYP3DKE4EfA3+SZNPlWZskSZI026xyAWdM3tG2/zzBmK2SHJzkA0neMsFzMZsBzwdurKrbhhzv3QK3S1/fK2gC1req6oH+wW1A+nr78VUT/RGSJEmSnmxVvEVtuSTZDngJTSAZeXsb8NcDnx9P8nng4Kp6pK//BW1744jz3NS2WyxnzUhJrhpxaKLnkyRJkqTOmY1XcA5s25NHHL8NOIgmhKwDPAf4c5rNCt4JnDIwfr22XTLifL3+9ZezRpIkSdIkZtUVnCTr0YSVkZsLVNXFwMV9XUuBLyf5DvB9YN8kf19V31/By52yqtpmWH97ZWfrlbwcSZIkacbMtis4bwbWZhk2F6iqnwLnth937DvUu9qyHsP1+u9bzhpJkiRJk5htAae3ucA/LWP9L9t2nb6+G9p21PMym7dt//M2y1IjSZIkaRKzJuAkeTnNC0JvrKqLlvE0L2/b/hdw3gL8BNgiySZDal7bthf09X0HeBjYIcnvDKzzacBr2o8TbYIgSZIkacCsCTj8dnOBibaGJsnLhvQ9LckHge2Au4Hze8fa9/Gc1H78hzag9Or2AF4JXEffcz1V9SDwRZorQfMHpnsvMBf4elXdiiRJkqQpW+U2GUjyBuAN7cdnte12SRa0v99dVYcN1DwD2Bt4FPiXSaa4IskPaTYUuIPmeZgdgBfTbDjwl1V1/0DNccDuwJ7Ad5N8k+bdOHu1NQf0vwC0dTiwM3Bokq2Ay4EtgT2Au4D3TLJOSZIkSQNWuYADbAXsN9C3afsD8GPgsIHjf0lztWThFDYXOBb4bzQv5twQeILmFrTPAMcNu6pSVY8meTXwAWBf4BDgfuBM4CNVdd2Qml+17+T5CE1geyXwK+BU4MNVdfsk65QkSZI0YJULOFU1n6fe1jVZzeeAz01x7PunvyqoqqXAh9ufqdbcQ/NC0cGXikqSJElaBrPpGRxJkiRJHWfAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZBhxJkiRJnWHAkSRJktQZq1zASbJnkhOSXJrk/iSV5LQRY+e2x0f9LJxgnv2SXJ7kwSRLklyUZPcJxs9JckiSa5M8nOSeJOcm2X6CmrWSHJXkhiSPJLkryZeSbDm9b0WSJEkSwGozvYBlcCTwR8CDwO3AC6dQ833gzCH9Pxw2OMmxwLz2/CcDqwP7AGcnOaiqThwYH2AhsCdwA3AisCGwN3BJkjdV1VkDNWsA3wB2AK4EPg08D9gLeF2SXarqu1P42yRJkiS1VsWAcwhN8LgZ2Am4cAo111TV/KmcvL3iMg+4Bdi2qu5t+48BrgKOTXJOVS3uK9uHJtwsAnatqkfampOAy4CTk1xQVQ/01RxKE27OAPauqifamtNpwtgpSV7S65ckSZI0uVXuFrWqurCqbqqqWkFTvKttP94LN+28i4HPAGsA+w/UvLttj+yFm7bmCuB04Jk0AQj4zRWf3jx/0x9i2is9lwJ/SBPgJEmSJE3RKhdwltFzkrwzyeFt+9IJxu7StucPOXbewBiSrAlsDyylCSaT1gCbAc8Hbqyq26ZYI0mSJGkSq+Itasvi1e3PbyS5CNivqn7S17cO8Fzgwaq6c8h5bmrbLfr6NgPmALdW1WNTrHlB2944Yr3DakZKctWIQ1N5PkmSJEnqjK5fwVkKfAzYBtig/ek9t7Mz8M021PSs17ZLRpyv17/+DNRIkiRJmkSnr+BU1V3Ahwe6L0nyGpqH/18OvJ1mB7NVVlVtM6y/vbKz9UpejiRJkjRjun4FZ6j2VrLPtx937DvUu3KyHsP1+u+bgRpJkiRJk5iVAaf1y7b9zS1qVfUQcAewbpJnD6nZvG37n525BXgc2DTJsCtiw2puaNtRz9gMq5EkSZI0idkccF7RtrcO9F/QtrsNqXntwBjabaEXAWsDr5xKDU0o+gmwRZJNplgjSZIkaRKdDjhJtk7ylL8xya40LwwFOG3g8Elte0SSDfpq5gLvAR4FTh2o+VzbHt1uG92r2RbYm+Zq0Vd6/e07fHrz/EP/GpPsQROUrgMunvSPlCRJkvQbq9wmA0neALyh/fistt0uyYL297ur6rD29+OAzZMsAm5v+17Kb98v86GqWtR//qpalOQ44FDg2iRnAKvTBJUNgYPal372Wwi8keZlnt9LcjawUVszB3hHVd0/UHMcsHtb890k36R5N85eNLu/HdD/AlBJkiRJk1vlAg6wFbDfQN+m7Q/Aj4FewPki8GfAtjS3fT0d+AXwJeDEqhr2Yk6qal6SH9BcsTkQeAK4Gjimqs4ZMr6S7Etzq9oBwEHAI8AlwNGDIaqteTTJq4EPAPvSXFG6HzgT+EhVXTf5VyFJkiSp3yoXcKpqPjB/imO/AHxhGedZACyYxvjHgE+2P1OtWUqzjfXgVtaSJEmSlkGnn8GRJEmSNLsYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmcYcCRJkiR1hgFHkiRJUmeMNeAkefo4zydJkiRJ0zHuKzh3JPn7JH8w5vNKkiRJ0qTGHXCeBrwfuCHJN5K8KcmcMc8hSZIkSUONO+A8B3gzcCmwK/Al4PYkH08yd8xzSZIkSdKTjDXgVNWvq+rfq2pn4IXAp4DVgA8CNyc5N8keSdzcQJIkSdLYrbCgUVU3VtU84Ln89qrObsD/An6SZH6S56yo+SVJkiTNPiv8SkpV/Rr4GvBV4GdAaG5l+zBwW5JPJVljRa9DkiRJUvet0ICT5BVJTqUJNp8E1gGOB7YCDgBuAA6iuZVtqufcM8kJSS5Ncn+SSnLaiLGbJ/nbJBck+WmSXyf5RZKzkrxqRM3b2nOO+nnXiLq1khyV5IYkjyS5K8mXkmw5wd+yYRvwFid5NMnPkpySZOOpfh+SJEmSfmu1cZ8wye8AbwHeCbyY5orN94DPAv9eVQ+3Q69N8kXgfGBP4N1TnOJI4I+AB4HbaZ71GeVjwN7AdcC5wD3AC4DXA69P8tdVdfyI2rOAa4b0XznY0V6B+gawQ3v808DzgL2A1yXZpaq+O1CzEbAI2AK4AFjY/i37tzXbVdWtE/xtkiRJkgaMNeAk+QLw58DawKPAF4HPVtXlw8ZX1eNJLgJ2mcY0h9AEm5uBnYALJxh7PvD3VfW9gXXuRBNIjkny5aq6c0jtmVW1YIprOpQm3JwB7F1VT7TznA6cCZyS5CW9/tbf0YSb49pnlXprex9NQPoszTNLkiRJkqZo3Leo7Q/8HPgbYOOqetuocNPnIuCjU52gqi6sqpuqqqYwdsFguGn7L27nXR3YfqpzD5MkQO+2tb/pDzFVdRbN5gp/SBPGejXr0lzlegiYP3DKE4EfA3+SZNPlWZskSZI024z7FrXdquo/plNQVd8CvjXmdUzFf7btYyOOb5XkYGBN4A7gwqq6fci4zYDnAzdW1W1Djp8HvJLmKlXvatMrgLWA/6iqB/oHV9UTSb4OHAi8Cpj0NrUkV404NNHte5IkSVLnjDXgTDfczJQk/xfNi0iXApeMGPbXA58fT/J54OCqeqSv/wVte+OI89zUtlssZ40kSZKkSYz1FrUku7a7gA19v02S57THdx7nvNPRbgjwb8AawPyqundgyG00O7u9gGbXt+fQPFe0mGbjhFMGxq/XtktGTNnrX385a0aqqm2G/QA/mkq9JEmS1BXjvkXtIOCFVfWzYQer6mdJtqP5B/5FY557Uknm0Gx8sANwOnDs4Jj2+ZyL+7qWAl9O8h3g+8C+Sf6+qr6/EpYsSZIkaRrGvcnA1jRbH0/kMuBlY553Um24OY1m6+YvAW+eykYFPVX1U5qtpgF27DvUu9qyHsP1+u9bzhpJkiRJkxh3wPk9mpd6TuQX7biVJsnTgf8J7AP8O/AXVTVqc4GJ/LJt1+nru6FtRz0vs3nb9j9vsyw1kiRJkiYx7oCzhOYFlxN5Hs32yCtFktWBL9NcuflX4C1V9fgynu7lbdu/s9ktwE+ALZJsMqTmtW17QV/fd4CHgR3aF6P2r/dpwGvajxO940eSJEnSgHEHnMuBNyR51rCD7eYDb2jHrXDthgJfBfYAvgDsP/CyzWE1T7l9LsnTknwQ2A64m+YFogC0t7md1H78hzag9Or2oNki+jr6nuupqgdpngVah6e+B+e9wFzg61U16RbRkiRJkn5r3JsMnAC8Drg0yTyaf6Q/2gaN3YB/BNYFjl/WCZK8gSYkAfSC1HZJFrS/311Vh7W/nwT8KU0ouQP4cPNezie5qKou6vt8RZIf0mwocAfN8zA7AC+m2XDgL6vq/oFzHAfsDuwJfDfJN2nejbNXW3PAkGB1OLAzcGiSrWhC35Y0Yewu4D2TfReSJEmSnmzs78FJ8jHgQzRXTirJvcAGQNqfj1XV+ROcZjJbAfsN9G3a/gD8GOgFnN4tY78LfHiCc17U9/uxwH+jeTHnhsATNLegfQY4bthVlTbEvRr4ALAvcAhwP3Am8JGqum5Iza/aHeU+QhPYXgn8CjgV+PCIl4pKkiRJmsC4r+BQVR9J8i2aLaNfTvMul3tonjs5oaq+sZznn89Tb+saNXbnZTj/+6db09YtpQlREwWpwZp7aF4oOvhSUUmSJEnLYOwBB5orOcB/rIhzS5IkSdIo495kQJIkSZJmzAq5ggOQZG2aZ2/mDDteVT9ZUXNLkiRJmp3GHnCSvAX4W5odwUapFTG3JEmSpNltrCEjyduAU4DHgUuBnwKPjXMOSZIkSRpl3FdRDgPuBf57VV0/5nNLkiRJ0oTGvcnAHwBfNtxIkiRJmgnjDjj3AI+O+ZySJEmSNCXjDjjnADsnyZjPK0mSJEmTGnfA+SCwBnBSknXHfG5JkiRJmtC4Nxn4MrAUeDvwF0luAu4bMq6qatcxzy1JkiRplht3wNm57/d1gK1GjKsxzytJkiRJ4w04VTXuW94kSZIkacoMJJIkSZI6w4AjSZIkqTPGHnCSPC3JQUm+k2RJksf6jv1xks8m2WLc80qSJEnSWANOktWBbwCfAjYDHgD634lzG3AA8JfjnFeSJEmSYPxXcN4PvAo4Cvh94PP9B6vqPuAS4E/GPK8kSZIkjT3g/CXwrar6aFU9wfDtoG8Dnj/meSVJkiRp7AFnE+A7k4y5B9hwzPNKkiRJ0tgDziPA+pOMeT5w35jnlSRJkqSxB5xrgNe0mw08RZL1aJ6/uXzM80qSJEnS2APOPwPPA/4tyTP6DyRZH1gAbACcNOZ5JUmSJInVxnmyqvqfSV4NvA14PXAvQJIrgRcBawCfqapzxzmvJEmSJMEKeNFnVR1A866b64Bn0rwHZ2vgZuCvquqgcc8pSZIkSTDmKzg9VbUAWJBkLZpb0pZU1UMrYi5JkiRJ6lkhAaenqh4GHl6Rc0iSJElSz9hvUZMkSZKkmTLWKzhJbp3i0KqqzcY5tyRJkiSN+xa1pwE1pH99YL32958B/znmeSVJkiRp7NtEzx11LMkfAMcD69C87FOSJEmSxmqlPYNTVTcDbwSeC3xkZc0rSZIkafZYqZsMVNUjwDeAfVfmvJIkSZJmh5nYRe0x4FkzMK8kSZKkjlupASfJ7wJ/Bvx0Zc4rSZIkaXYY9zbRH55gnucBe9DspvbB5ZhjT2AnYCvgj4DfAf6tqt48Qc32wJHAK4C1gJuAU4ATqurxETW7A4cBfwzMAf438Nmq+pcJ5tkPeA/wh8DjwPeAY6vqnBHj5wDvA/YHNqd5Kep3gKOratGoeSRJkiQNN+5toudPcvx+mn+8/8NyzHEkTbB5ELgdeOFEg5PsAXwFeAQ4HbgH+B/AJ4EdgL2G1LwXOAH4FXAa8GtgT2BBkpdU1WFDao4F5rVrOhlYHdgHODvJQVV14sD4AAvb894AnAhsCOwNXJLkTVV11hS+D0mSJEmtcQecV43ofwK4F/hRVT22nHMcQhMibqa5knPhqIFJnkETNh4Hdq6qK9v+DwEXAHsm2aeqFvbVzAWOpQlCL6uqxW3/R4ErgHlJvlJV3+6r2Z4m3NwCbFtV97b9xwBXAccmOad3rtY+NOFmEbBruwEDSU4CLgNOTnJBVT2wDN+RJEmSNCuN9Rmcqrp4xM+lVfXDMYQbqurCqrqpqoa9UHTQnsAzgYW9cNOe4xGaK0EA7x6oOQBYAzixP5C0oeXv2o/vGqjpff54L9y0NYuBz7Tn23+gpjfvkb1w09ZcQXOl6Znt+iVJkiRN0UzsorYy7dK25w85dgmwFNg+yRpTrDlvYMwy1SRZE9i+nf/SacwjSZIkaQLj3mTg+ctaW1U/GedaWi9o2xuHzPdYktuAFwGbAtdPoebOJA8BGydZu6qWJlmH5uWlD1bVnUPWcFPbbtHXtxnNxgW3jriqNaxmpCRXjTg04fNJkiRJUteM+xmcxcBUbh0bVIx/LdDs2AawZMTxXv/606xZpx23dAXOMVgjSZIkaRLjDhX/CswFdqT5R/o1wM9pXuy5Fc0/7C+mCUIak6raZlh/e2Vn65W8HEmSJGnGjDvg/L/At2m2YD6qqu7vHWh3NDsKeCvwzqp6yi1gK0DvSsh6I473+u8bqPnd9tivJqhZMtBOd47p1kiSJEmaxLg3GfgE8IOqmtcfbgCq6v6qOoTmhZmfGPO8o9zQtk95liXJasAmwGPArVOseTbN7Wm3V9VSgKp6CLgDWLc9Pmjztu0PdLfQbF29abuOqdRIkiRJmsS4A86ONO9wmchlNO+vWRkuaNvdhhzbEVgbWFRVj06x5rUDY5appt0WelE7/yunMY8kSZKkCYw74KxB87zNRJ7djlsZzgDuBvZJ8rJeZ7tN89Htx88N1JwKPAq8t33pZ69mA+Dw9uNJAzW9z0e043o1c4H3tOc7daCmN+/R7Xp6NdsCewO/BL4yyd8nSZIkqc+4n8H5Hk2YOLGqvjd4MMk2NP94v/IplVOU5A3AG9qPvTC1XZIF7e93V9Vh0NwWl+QdNEHnoiQLgXuA19NsB30GzUs1f6OqbkvyfuB44MokpwO/pnnp5sbAP1bVtwdqFiU5DjgUuDbJGcDq7d+6IXBQ/0tDWwuBN7bn/V6Ss4GN2po5wDsGb/OTJEmSNLFxB5yjaF52+Z0k/0bzMs1fAL9Pc1vaX9BcNTpqOebYCthvoG/T9gfgx8BhvQNVdWaSnYAjgDcBawI304SR46vqKdtaV9UJSRa353lru+brgCOr6l+GLaqq5iX5Ac0VmwOBJ4CrgWOq6pwh4yvJvjS3qh0AHAQ8QvOdHV1Viyb/KiRJkiT1G2vAqar/P8k+wD8Bb+PJQSTAvcCBVfXN5ZhjPjB/mjXfAv50mjVnA2dPs2YBsGAa4x+j2XHuk9OZR5IkSdJwY3+5ZlWdkeQ8YA+ad7CsR7Mt8tXAWe2uY5IkSZI0dmMPOPCbrZP/vf2RJEmSpJVi3LuoPUmSDZI8b0XOIUmSJEk9Yw84SdZN8o9Jfk6zRfNtfcdenoOxkNkAABzvSURBVOTcJFuPe15JkiRJGmvASbIe8G3gEOBnwPU0mwv0/IDmxZb7jnNeSZIkSYLxX8E5AngR8Laq2hr4cv/BqloKXAzsOuZ5JUmSJGnsAeeNwNer6l8nGPNj4LljnleSJEmSxh5wNgaunWTMgzRbR0uSJEnSWI074DwA/N4kYzah2XxAkiRJksZq3AHnCmD3JL8z7GCSZwN/Clw25nklSZIkaewB59PARsC5SbbsP9B+/jKwJnD8mOeVJEmSJFYb58mq6utJjgI+AvwQ+E+AJHcDG9BsGf23VbVonPNKkiRJEqyAF31W1VE020D/f8C9wONAAecC/3dVHTPuOSVJkiQJxnwFJ8mOwP1VdSFw4TjPLUmSJEmTGfcVnAuBA8d8TkmSJEmaknEHnLuBh8d8TkmSJEmaknEHnIuA7cd8TkmSJEmaknEHnCOBFyT5WJKnj/nckiRJkjShsW4yAHyQZnvow4G/SvJ94Oc0u6j1q6r6qzHPLUmSJGmWG3fAeVvf789qf4YpwIAjSZIkaazGHXA2GfP5JEmSJGnKljvgJHkrcE1VXVtVPx7DmiRJkiRpmYxjk4EFwBv6O5Lsl+SCMZxbkiRJkqZs3Luo9cwFdlpB55YkSZKkoVZUwJEkSZKklc6AI0mSJKkzDDiSJEmSOmNcAWfwRZ6SJEmStNKN6z0485PMH+xM8viI8VVV434HjyRJkqRZblwhIyt4vCRJkiRNarkDTlX5HI8kSZKk/xIMJ5IkSZI6w4AjSZIkqTMMOJIkSZI6w4AjSZIkqTMMOJIkSZI6o/MBJ8nbktQkP4/3jZ87ydiFE8y1X5LLkzyYZEmSi5LsPsH4OUkOSXJtkoeT3JPk3CTbj/t7kCRJkmaD2fCyzWuAo0YceyWwC3DekGPfB84c0v/DYSdKciwwD7gdOBlYHdgHODvJQVV14sD4AAuBPYEbgBOBDYG9gUuSvKmqzpr4T5MkSZLUr/MBp6quoQk5T5Hk2+2v/zzk8DVVNX8qc7RXXOYBtwDbVtW9bf8xwFXAsUnOqarFfWX70ISbRcCuVfVIW3MScBlwcpILquqBqaxBkiRJ0iy4RW2UJC8BXgHcAXxtOU/3rrb9eC/cALSB5jPAGsD+AzXvbtsje+GmrbkCOB14Jk0AkiRJkjRFszbgAAe27Req6vEhx5+T5J1JDm/bl05wrl3a9vwhx84bGEOSNYHtgaXApVOpkSRJkjS5zt+iNkyStYA3A48Dnx8x7NXtT3/dRcB+VfWTvr51gOcCD1bVnUPOc1PbbtHXtxkwB7i1qh6bYs1ISa4aceiFU6mXJEmSumK2XsH5c2B94Pyq+unAsaXAx4BtgA3an52AC4GdgW+2oaZnvbZdMmKuXv/6y1kjSZIkaRKz8goOv7097Z8GD1TVXcCHB7ovSfIamof/Xw68Hfj0Cl3hNFTVNsP62ys7W6/k5UiSJEkzZtZdwUnyIprnX24Hzp1qXXsrWe92th37DvWutqzHcL3++5azRpIkSdIkZl3AYfLNBSbyy7b9zS1qVfUQzU5s6yZ59pCazdv2xr6+W2ie/9k0ybCraMNqJEmSJE1iVgWcdveyt9CEiy8swyle0ba3DvRf0La7Dal57cAY2m2hFwFr07xsdNIaSZIkSZObVQEH2Itm04DzhmwuAECSrZM85XtJsitwSPvxtIHDJ7XtEUk26KuZC7wHeBQ4daDmc217dBu8ejXbAnvTXC36yuR/kiRJkqSe2bbJQO/2tH+eYMxxwOZJFtE8pwPwUn77TpoPVdWi/oKqWpTkOOBQ4NokZwCr0wSVDYGD2pd+9lsIvJHmZZ7fS3I2sFFbMwd4R1XdP/0/UZIkSZq9Zk3ASbIl8N+ZfHOBLwJ/BmxLc6vY04FfAF8CTqyqYS/mpKrmJfkBzRWbA4EngKuBY6rqnCHjK8m+NLeqHQAcBDwCXAIcPRiiJEmSJE1u1gScqroeyBTGfYFlez6HqloALJjG+MeAT7Y/kiRJkpbTbHsGR5IkSVKHGXAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdcasCDhJFiepET8/H1GzfZJzk9yT5OEk1yY5OMmcCebZPclFSZYkeTDJd5PsN8na9ktyeTt+SVu/+/L+zZIkSdJstNpML2AlWgJ8akj/g4MdSfYAvgI8ApwO3AP8D+CTwA7AXkNq3gucAPwKOA34NbAnsCDJS6rqsCE1xwLzgNuBk4HVgX2As5McVFUnTv/PlCRJkmav2RRw7quq+ZMNSvIMmrDxOLBzVV3Z9n8IuADYM8k+VbWwr2YucCxNEHpZVS1u+z8KXAHMS/KVqvp2X832NOHmFmDbqrq37T8GuAo4Nsk5vXNJkiRJmtysuEVtmvYEngks7IUbgKp6BDiy/fjugZoDgDWAE/sDSRta/q79+K6Bmt7nj/fCTVuzGPhMe779l+cPkSRJkmab2XQFZ40kbwaeDzwEXAtcUlWPD4zbpW3PH3KOS4ClwPZJ1qiqR6dQc97AmKnMcx7woXbMR4Ycf5IkV4049MLJaiVJkqQumU0B51nAFwf6bkuyf1Vd3Nf3gra9cfAEVfVYktuAFwGbAtdPoebOJA8BGydZu6qWJlkHeC7wYFXdOWStN7XtFlP5wyRJkiQ1ZkvAORW4FPjfwAM04eS9wIHAeUm2q6rvt2PXa9slI87V61+/r28qNeu045Yu4xwjVdU2w/rbKztbT+UckiRJUhfMioBTVUcNdP0QeFeSB2ke9J8P/NnKXpckSZKk8Zrtmwyc1LY79vX1rp6sx3C9/vuWoWbJQDudOSRJkiRNYrYHnF+27Tp9fTe07VOef0myGrAJ8Bhw6xRrnt2e//aqWgpQVQ8BdwDrtscHbd62T3mmR5IkSdJosz3gvKJt+8PKBW2725DxOwJrA4v6dlCbrOa1A2OWp0aSJEnSBDofcJJs2e5aNtg/Fzix/Xha36EzgLuBfZK8rG/8msDR7cfPDZzuVOBR4L3teXs1GwCHtx9PGqjpfT6iHde/rve05zt1gj9NkiRJ0oDZsMnA3sC8JJcAP6bZRW0z4HXAmsC5wLG9wVV1f5J30ASdi5IsBO4BXk+zHfQZwOn9E1TVbUneDxwPXJnkdODXNC8N3Rj4x6r69kDNoiTHAYcC1yY5A1i9Xe+GwEH9Lw2VJEmSNLnZEHAupAkmfwzsQPM8zH3AZTTvxfliVVV/QVWdmWQn4AjgTTRB6GaaMHL84Pi25oQki4HDgLfSXB27Djiyqv5l2MKqal6SH9BcsTkQeAK4Gjimqs5Zzr9bkiRJmnU6H3Dal3hePOnAp9Z9C/jTadacDZw9zZoFwILp1EiSJEkarvPP4EiSJEmaPQw4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMww4kiRJkjrDgCNJkiSpMzofcJJslOTtSb6a5OYkDydZkuSyJH+V5GkD4+cmqQl+Fk4w135JLk/yYDvHRUl2n2D8nCSHJLm2Xdc9Sc5Nsv04vwNJkiRptlhtphewEuwFfA64E7gQ+Anw+8Abgc8Dr02yV1XVQN33gTOHnO+HwyZJciwwD7gdOBlYHdgHODvJQVV14sD4AAuBPYEbgBOBDYG9gUuSvKmqzpr+nytJkiTNXrMh4NwIvB74WlU90etMcjhwOfAmmrDzlYG6a6pq/lQmaK+4zANuAbatqnvb/mOAq4Bjk5xTVYv7yvahCTeLgF2r6pG25iTgMuDkJBdU1QPT+3MlSZKk2avzt6hV1QVVdXZ/uGn7fw6c1H7ceTmneVfbfrwXbto5FgOfAdYA9h+oeXfbHtkLN23NFcDpwDNpApAkSZKkKep8wJnEf7btY0OOPSfJO5Mc3rYvneA8u7Tt+UOOnTcwhiRrAtsDS4FLp1IjSZIkaXKz4Ra1oZKsBry1/TgsmLy6/emvuQjYr6p+0te3DvBc4MGqunPIeW5q2y36+jYD5gC3VtWwcDWsZqQkV4049MKp1EuSJEldMZuv4HwCeDFwblV9va9/KfAxYBtgg/ZnJ5oNCnYGvtmGmp712nbJiHl6/esvZ40kSZKkSczKKzhJ3kezKcCPgLf0H6uqu4APD5RckuQ1NA//vxx4O/DplbDUKamqbYb1t1d2tl7Jy5EkSZJmzKy7gpPkvTTh5DrgVVV1z1Tq2lvJPt9+3LHvUO9qy3oM1+u/bzlrJEmSJE1iVgWcJAcDJ9C8y+ZV7U5q0/HLtv3NLWpV9RBwB7BukmcPqdm8bW/s67sFeBzYtH0WaCo1kiRJkiYxawJOkr8FPglcQxNu7lqG07yibW8d6L+gbXcbUvPagTG020IvAtYGXjmVGkmSJEmTmxUBJ8mHaDYVuIrmpZp3TzB26yRP+V6S7Aoc0n48beBw7306RyTZoK9mLvAe4FHg1IGaz7Xt0e220b2abYG9aa4WDb58VJIkSdIEOr/JQJL9gI/S3BJ2KfC+JIPDFlfVgvb344DNkywCbm/7Xspv30nzoapa1F9cVYuSHAccClyb5AxgdZqgsiFwUPvSz34LgTfSvMzze0nOBjZqa+YA76iq+5f175YkSZJmo84HHGCTtp0DHDxizMXAgvb3LwJ/BmxLc6vY04FfAF8CTqyqYS/mpKrmJfkBzRWbA4EngKuBY6rqnCHjK8m+NLeqHQAcBDwCXAIcPRiiJEmSJE2u8wGnquYD86cx/gvAF5ZxrgX8NihNZfxjNM8FfXJZ5pMkSZL0ZLPiGRxJkiRJs4MBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHBmWJKNk5yS5GdJHk2yOMmnkmww02uTJEmSVjWrzfQCZrMkmwGLgN8DzgJ+BPw34K+B3ZLsUFW/msElSpIkSasUA87M+ixNuHlfVZ3Q60xyHHAI8HHgXTO0Nk3T3A98baaXsNIt/sTrZnoJkiRJT+ItajOkvXrzGmAx8JmBwx8BHgLekmSdlbw0SZIkaZXlFZyZ86q2/Y+qeqL/QFU9kORbNAHoFcA3V/bipKnwqpUkSfqvxoAzc17QtjeOOH4TTcDZgkkCTpKrRhz6o+uvv55tttlm2Va4HO68Y8lKn1NaGdZY8NczvQRJY/Di564300vQSvDDWfjvkZn4b/v6668HmLvSJx7BgDNzev/1jfr/vF7/+ssxx+MPP/zwkquvvnrxcpxjul7Ytj9aiXPONn7HK57f8Yrnd7zi+R1P4OpfjOU0fscrlt/vMpjmf9vj+o7nAvcv5znGxoDTAVW18i/RjNC7mvRfaU1d43e84vkdr3h+xyue3/GK53e8Yvn9rnhd/Y7dZGDm9K7QjLqO2Ou/byWsRZIkSeoEA87MuaFttxhxfPO2HfWMjiRJkqQBBpyZc2HbvibJk/7vkOR3gB2ApcB3VvbCJEmSpFWVAWeGVNUtwH/QPJT1noHDRwHrAF+sqodW8tIkSZKkVZabDMys/wdYBByfZFfgeuDlNO/IuRE4YgbXJkmSJK1yUlUzvYZZLcnzgI8CuwEbAXcCXwWOqqp7Z3JtkiRJ0qrGgCNJkiSpM3wGR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR5IkSVJnGHAkSZIkdYYBR2ORZOMkpyT5WZJHkyxO8qkkG8z02lak9u+sET8/H1GzfZJzk9yT5OEk1yY5OMmcCebZPclFSZYkeTDJd5PsN8na9ktyeTt+SVu/+wTj5yQ5pF3Pw+36zk2y/dS/kWWXZM8kJyS5NMn97Xd42iQ1nfkuk6yV5KgkNyR5JMldSb6UZMuJ1jYd0/mOk8yd4L/tSrJwgnn+S35fSTZs/3dpcfu/Uz/7P+2da6wdVRXHfwsaW1qglvqhYjG3VGoUlAQxYCvY1thoYlOiftBEKQgJfChYYxMjUSkYQwtFNFiCRhBfGFMS+kGhlfQpJTUYaQQfpS19qU1LkTbtvbZQXH7Y65aTYc7pmfOYc+70/0t25s6avfbe87/r7DN7Zs7e0W9NrudTBDObaGY3mtnjZrY9zuWwmT1tZjeYWe53ruK4eYpqrDgujpktNbM1Zra35jyeM7PbzWxiHR/FcAGKaKwYbhF3V1JqKwFTgf2AAyuBJcDa2P8HMLHXbeziue8CDgGLc9KinPzzgBPAUeAh4J7QyIEVdepYEMcPAsuB+4C9YVtWx2dZHN8b+ZcDr4RtQU5+A1bU/M/uifYdjfbOK0HLLVH/EeDv8fcvG+SvjJbAaODp8HkWWAo8CrwODAJXlK0xMBDHt9SJ78+NJL1ICylvDZ81pH5qZezvBy7sgL43R3n/Bn4F3AU8TOojHHiMWH9OcVyOxorjljR+Ddgcui4B7o/2OfAv4ALFcHkaK4Zb1LgThSid3glYHYF5S8b+vbA/2Os2dvHcdwG7msx7LnAAOA5cXmMfAzwTWn0+4zMAHItOaaDGPgHYHj4fyfhMD/t2YEKmrFeivIGMzxfCZxMwpsb+4WjvAeCcLms5C7goOtmZNL74rpSWwDfCZwVwRo19Xtj/WmsvSeOBOP5IgfL7Vi/gR3Hs3oz91rCv6oC+s4G5OXVPAvZEPZ9VHJeqseK4uMZj6ti/G3U8oBguVWPFcCsad6IQpdM3kZ7eOLAzJ4jPIY38B4FxvW5rl85/F80PcL4cWv0s59jsOLYhY78z7Hc0Wx7w87Bfn+OTWx6wMeyzcnzqltdFXWfS+OK7MlqSBhu7wz4lx6dueV3WeIDiX6p9qRdwNjAU/VH2C/qM+Bw7HbpzWEeb26KO+xXHpWqsOO6cvpdG+U8phrsWw3kaK4ZbSPoNjmiXWbH9vbv/r/aAux8h3QkYC1xZdsNKZLSZfdHMbjOzr5jZrDrvHc+O7aqcYxtJH/rpZja6SZ8nM3la8jGzMaS7PUPAHwrU00uqpOVU4N3Ai+6+s0DbyuJ8M7sp4vsmM/tgg7z9qteVwFnApuiXThL91urYnUX3eD22J2psiuPOkqfxMIrj9pkb27/U2BTDnSVP42EUwwUY1W4B4rTnvbF9sc7xbcAcYBrpXcsqMgn4Rca208yud/cNNba6Wrn7CTPbCVwMXEj6fcSpfPaZ2SAw2czGuvuQmY0D3gUcdfd9OW3dFttpNbapwJnAS+6ed2GQ59NrqqRlM5+hrE+ZfCLSScxsPTDf3ffU2PpZr55qbGajgGtjt/aCQ3HcIRpoPIziuCBmtoh0x308cDnwUdKF95Jm2qQYPjVNajyMYrgAeoIj2mV8bA/XOT5sf3sJbekFPwU+ThrkjAM+QHq/dAB40swurcnbilbN+ozPbLtRRz/9D6ukZb/qPwR8B/gQ6d34CcDHgHWk19vWxBfpMP2sV681XgJcAjzh7qtr7IrjzlFPY8Vx6ywCbgcWki68VwFz3P3lNtukGH6TZjRWDLeABjhCtIG73+Hua919v7sPufsL7n4zaYKFs0gznAgx4nD3A+7+bXf/s7sfirSR9ET2j8B7gBt728r+x8xuBb5GmpXoSz1uTiVppLHiuHXcfZK7G+kG3mdIT2GeM7PLetuy6tCMxorh1tAAR7RL9k5LlmH7oRLa0k88GNura2ytaNWsz+HMtht19NP/sEpajij94/WFn8RuWfE9IjU2swXAD4C/kX5o+59MFsVxmzShcS6K4+aJG3iPky6oJ5J+cN5OmxTDGU6hcT0fxXADNMAR7bI1tvXel7wotvXet6wqw4+Xax8b19Uq3h+fQvpx7EtN+rwzyv+nuw8BuPsgaQ79s+N4lrz/xw7gDeDCaEczPr2mSlqOxM/QW+K7z/UqXWMzW0ha2+IF0oV33sK/iuM2aFLjRiiOC+Duu0kDyYvN7B2napNiuDh1NG6EYrgOGuCIdlkX2zn21hWkzwFmkN4f3Vx2w3rM8KxxtZ362th+Mif/1aTZ5p5x9+NN+nwqk6clH3c/RlqvYCxwVYF6ekmVtNxBWr9jmplNKdC2XpIX39C/em0G/gvMiH7pJNFvzYnddXQAM/s6aWG9LaQL7wN1siqOW6SAxo1QHBfn/Ni+kWmfYrhzZDVuhGK4HkXnlVZSyiZO04U+gfeRs74PaYKBbXHut9XYzyXdbSmyINoU+mdBtHNL1HYmjddoqZSWlLS4XEGNL8urkzSpxrHwnT5S9KKsxeXgW1Hen4DzTpFXcdx9jRXHxbSdBozPsZ/Bm4tQblIMl6qxYrgVnTtRiNLpnUjTC+6PwFwJ3EUasTvpceTEXrexS+e9GDgC/A54AFgKPEa6O+Fhf1vG5xrS4/qjpHdn7yb9MHa4g7Ccem6J4weB5aS7lnvDtqxO2+6N43sj//Lwd2BBTn6L+p00lefdwEPRzhPAvBL0vAZ4JNKqaMuOGtuynPyV0BIYTfpSceBZ0oxQj5LW9RgErihbY2A96TWHFXHe95GmevdI3xxJepHea98aPmtI/dTK2N8PTO2AvvOjvBNx7otz0nWK4/I0RnFcVN+FpO+wp4AfR/kPk/oJB/YB71cMl6exYrhFnTtRiJIScAFpyuR9wGuk1W2/T82dg6ol0jSNv46O/FB8oF+OTuvavE49/GYATwCvRif3PPBV4MwGdc0FNpAGVIPRkcw/Rfuui3yD4bcB+HSD/KOiHc9Hu16Ndk5vVE8H9Vxc02HnpV1V1pL0asCdpKd/xyOWVpC5mChLY+AG4LeklaWPRpv2AL8BrhqJegHnkX6QvpvUT+0jXVhMLklfB9YrjsvTWHFcWN9LgB+SXv07SLqwPRw6LKbOEzPFcPc0Vgy3liwqEkIIIYQQQogRjyYZEEIIIYQQQlQGDXCEEEIIIYQQlUEDHCGEEEIIIURl0ABHCCGEEEIIURk0wBFCCCGEEEJUBg1whBBCCCGEEJVBAxwhhBBCCCFEZdAARwghhBBCCFEZNMARQgghhBBCVAYNcIQQQgghhBCVQQMcIYQQQgghRGXQAEcIIYQQQghRGTTAEUIIIYQQQlQGDXCEEEIIIYQQlUEDHCGEEEIIIURl0ABHCCGEEEIIURk0wBFCCCGEEEJUhv8Dou2Uwc7JhqIAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "image/png": {
       "height": 248,
       "width": 412
      },
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "loci_df.SUM_DEPTH.plot(kind=\"hist\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.Fis == -9])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(17187, 20783)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.SUM_DEPTH <5000]),len(loci_df[loci_df.SUM_DEPTH >=500])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1858, 18717)"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.MEAN_DEPTH > 30]),len(loci_df[loci_df.MEAN_DEPTH >= 3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "273"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL >= 10]) - len(loci_df[loci_df.QUAL >= 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(8283, 7239, 5813)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(loci_df[loci_df.QUAL <  200]), len(loci_df[loci_df.QUAL < 100]), len(loci_df[loci_df.QUAL < 20])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2217, 48)"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    " len(loci_df[loci_df.Fis <= -0.5]), len(loci_df[loci_df.MAF < 0.02])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_snps(df, imputed=False):\n",
    "    if imputed:\n",
    "        return df[(df.MAF >= 0.01) &  \n",
    "                  (df.Fis > -0.5)]\n",
    "    else:\n",
    "        return df[(df.MEAN_DEPTH >= 3) & \n",
    "                  (df.MEAN_DEPTH < 25) & \n",
    "                  (df.QUAL >= 50) & \n",
    "                  (df.MAF >= 0.02) &  \n",
    "                  (df.Fis > -0.5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10067, 25)"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "loci_stage1 = filter_snps(loci_df)\n",
    "loci_stage1.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"stage1_positions.txt\"), \"w\") as o:\n",
    "    for elem in loci_stage1.index:\n",
    "        o.write(\"%s\\n\" % \"\\t\".join(elem.split(\"-\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "run_control": {
     "marked": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/snps_indv_removed.recode.vcf.gz\n",
      "\t--recode-INFO-all\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps\n",
      "\t--positions /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/stage1_positions.txt\n",
      "\t--recode\n",
      "\t--remove-filtered-all\n",
      "\t--remove-indels\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 10067 out of a possible 20784 Sites\n",
      "Run Time = 11.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!$vcftools --gzvcf $vcf_bad_remove_gz \\\n",
    "--remove-indels  \\\n",
    "--remove-filtered-all \\\n",
    "--recode \\\n",
    "--recode-INFO-all \\\n",
    "--positions {os.path.join(analysis_dir, \"stage1_positions.txt\")} \\\n",
    "--out {os.path.join(analysis_dir, \"good_snps\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "snps_gz = snps + \".gz\"\n",
    "!$bgzip -c {snps} > {snps_gz}\n",
    "!$tabix {snps_gz}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plink / chrom_map"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20784"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "chroms = sorted(set([x.split(\"-\")[0] for x in loci_df.index]))\n",
    "len(chroms) # number of unique contigs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(os.path.join(analysis_dir, \"chrom_map.txt\"), \"w\") as o:\n",
    "    for i, c in enumerate(chroms):\n",
    "        o.write(\"%s\\t%d\\n\" % (c, i))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "def write_plink_files(vcf_gz):\n",
    "    !$vcftools --gzvcf {vcf_gz} \\\n",
    "    --out {vcf_gz} \\\n",
    "    --plink \\\n",
    "    --chrom-map {os.path.join(analysis_dir, \"chrom_map.txt\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf.gz\n",
      "\t--chrom-map /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/chrom_map.txt\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf.gz\n",
      "\t--plink\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Writing PLINK PED and MAP files ... \n",
      "\tRead 20785 chromosome mapping file entries.\n",
      "Done.\n",
      "After filtering, kept 10067 out of a possible 10067 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "write_plink_files(snps_gz)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Make 012"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf\n",
      "\t--012\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Writing 012 matrix files ... Done.\n",
      "After filtering, kept 10067 out of a possible 10067 Sites\n",
      "Run Time = 1.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--012"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Get coverage per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - 0.1.16\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--gzvcf /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf\n",
      "\t--depth\n",
      "\t--out /data/gpfs/home/tfaske/d/frogs/ddocent/filtering/good_snps.recode.vcf\n",
      "\n",
      "Using zlib version: 1.2.11\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "Warning: Expected at least 2 parts in INFO entry: ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "Warning: Expected at least 2 parts in FORMAT entry: ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "After filtering, kept 297 out of 297 Individuals\n",
      "Outputting Mean Depth by Individual\n",
      "After filtering, kept 10067 out of a possible 10067 Sites\n",
      "Run Time = 0.00 seconds\n"
     ]
    }
   ],
   "source": [
    "f = os.path.join(analysis_dir, \"good_snps.recode.vcf\")\n",
    "!$vcftools --gzvcf {f} \\\n",
    "--out {f} \\\n",
    "--depth"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>INDV</th>\n",
       "      <th>N_SITES</th>\n",
       "      <th>MEAN_DEPTH</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>RL_AD1C_200</td>\n",
       "      <td>9610</td>\n",
       "      <td>6.56431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>RL_AD1C_201</td>\n",
       "      <td>9924</td>\n",
       "      <td>9.74688</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RL_AD1C_202</td>\n",
       "      <td>9949</td>\n",
       "      <td>12.52700</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RL_AD2C_203</td>\n",
       "      <td>9699</td>\n",
       "      <td>9.83967</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>RL_AD2C_204</td>\n",
       "      <td>9960</td>\n",
       "      <td>11.83330</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          INDV  N_SITES  MEAN_DEPTH\n",
       "0  RL_AD1C_200     9610     6.56431\n",
       "1  RL_AD1C_201     9924     9.74688\n",
       "2  RL_AD1C_202     9949    12.52700\n",
       "3  RL_AD2C_203     9699     9.83967\n",
       "4  RL_AD2C_204     9960    11.83330"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_file = os.path.join(analysis_dir, \"good_snps.recode.vcf.idepth\")\n",
    "depth_df = pd.read_csv(depth_file, sep=\"\\t\")\n",
    "depth_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "count    297.000000\n",
       "mean       8.719475\n",
       "std        2.710802\n",
       "min        2.138540\n",
       "25%        7.020650\n",
       "50%        9.222270\n",
       "75%       10.866300\n",
       "max       16.768300\n",
       "Name: MEAN_DEPTH, dtype: float64"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "depth_df.MEAN_DEPTH.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Remove unnecessary files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm snps*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *miss*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "!rm *vcf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
